{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scripts.pipeline_steps import step5 as s5\n",
    "from scripts.notebook_operations import setup_notebook\n",
    "\n",
    "engine, g37, transcripts, genes, mirnas = setup_notebook()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_87132/4074695647.py:13: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df[['in_cgc', 'is_driver']] = df[['in_cgc', 'is_driver']].fillna(False).infer_objects(copy=False)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>wt_prediction</th>\n",
       "      <th>mut_prediction</th>\n",
       "      <th>pred_difference</th>\n",
       "      <th>vcf_id</th>\n",
       "      <th>mirna_accession</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mutation_context</th>\n",
       "      <th>mutsig</th>\n",
       "      <th>is_intron</th>\n",
       "      <th>is_gain</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>in_cgc</th>\n",
       "      <th>is_driver</th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>seed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0000265</td>\n",
       "      <td>0.411319</td>\n",
       "      <td>0.724802</td>\n",
       "      <td>0.313</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0000265</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-204-5p</td>\n",
       "      <td>miR-204-5p/211-5p</td>\n",
       "      <td>GGCATAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0003301</td>\n",
       "      <td>0.289311</td>\n",
       "      <td>0.662916</td>\n",
       "      <td>0.374</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0003301</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-33b-5p</td>\n",
       "      <td>miR-33-5p</td>\n",
       "      <td>CAATGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0004517</td>\n",
       "      <td>0.351466</td>\n",
       "      <td>0.623310</td>\n",
       "      <td>0.272</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0004517</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-106a-3p</td>\n",
       "      <td>miR-106a-3p</td>\n",
       "      <td>TAAGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0005882</td>\n",
       "      <td>0.550857</td>\n",
       "      <td>0.279106</td>\n",
       "      <td>-0.272</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0005882</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-548k</td>\n",
       "      <td>miR-548av-5p/548k/8054</td>\n",
       "      <td>GCAAAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0005933</td>\n",
       "      <td>0.650570</td>\n",
       "      <td>0.375474</td>\n",
       "      <td>-0.275</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0005933</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-1277-3p</td>\n",
       "      <td>miR-1277-3p</td>\n",
       "      <td>AAATACA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    id  wt_prediction  mut_prediction  \\\n",
       "0  PD4005a_1_12844091_G_T_MIMAT0000265       0.411319        0.724802   \n",
       "1  PD4005a_1_12844091_G_T_MIMAT0003301       0.289311        0.662916   \n",
       "2  PD4005a_1_12844091_G_T_MIMAT0004517       0.351466        0.623310   \n",
       "3  PD4005a_1_12844091_G_T_MIMAT0005882       0.550857        0.279106   \n",
       "4  PD4005a_1_12844091_G_T_MIMAT0005933       0.650570        0.375474   \n",
       "\n",
       "   pred_difference   vcf_id mirna_accession    gene_id mutation_context  \\\n",
       "0            0.313  PD4005a    MIMAT0000265  not_found          A[C>A]A   \n",
       "1            0.374  PD4005a    MIMAT0003301  not_found          A[C>A]A   \n",
       "2            0.272  PD4005a    MIMAT0004517  not_found          A[C>A]A   \n",
       "3           -0.272  PD4005a    MIMAT0005882  not_found          A[C>A]A   \n",
       "4           -0.275  PD4005a    MIMAT0005933  not_found          A[C>A]A   \n",
       "\n",
       "  mutsig  is_intron  is_gain  gene_name gene_biotype  in_cgc  is_driver  \\\n",
       "0   SBS3      False     True  not_found    not_found   False      False   \n",
       "1   SBS3      False     True  not_found    not_found   False      False   \n",
       "2   SBS3      False     True  not_found    not_found   False      False   \n",
       "3   SBS3      False    False  not_found    not_found   False      False   \n",
       "4   SBS3      False    False  not_found    not_found   False      False   \n",
       "\n",
       "        mirna_name            mirna_family     seed  \n",
       "0   hsa-miR-204-5p       miR-204-5p/211-5p  GGCATAG  \n",
       "1   hsa-miR-33b-5p               miR-33-5p  CAATGCA  \n",
       "2  hsa-miR-106a-3p             miR-106a-3p  TAAGAAG  \n",
       "3     hsa-miR-548k  miR-548av-5p/548k/8054  GCAAAAT  \n",
       "4  hsa-miR-1277-3p             miR-1277-3p  AAATACA  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_threshold, wt_threshold = (0.5 , 0.5)\n",
    "df = pd.read_parquet('data/predictions.parquet')\n",
    "\n",
    "\n",
    "\n",
    "df = s5.filter_rows_by_thresholds(df, mut_threshold, wt_threshold)\n",
    "\n",
    "df[\"is_gain\"] = df.mut_prediction > df.wt_prediction\n",
    "\n",
    "df = df.merge(genes[[\"gene_id\", \"gene_name\", \"gene_biotype\", \"in_cgc\", \"is_driver\"]], on=\"gene_id\", how=\"left\")\n",
    "df = df.merge(mirnas[[\"mirna_accession\", \"mirna_name\", \"mirna_family\",  \"seed\"]], on=\"mirna_accession\", how=\"left\")\n",
    "\n",
    "df[['in_cgc', 'is_driver']] = df[['in_cgc', 'is_driver']].fillna(False).infer_objects(copy=False)\n",
    "df[['gene_id', 'mutsig', \"gene_name\", \"gene_biotype\"]] = df[['gene_id', 'mutsig', \"gene_name\", \"gene_biotype\"]].fillna(\"not_found\").infer_objects(copy=False)\n",
    "\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>in_cgc</th>\n",
       "      <th>is_driver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>1</td>\n",
       "      <td>182696</td>\n",
       "      <td>184174</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000228037</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>2581560</td>\n",
       "      <td>2584533</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000284616</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>5301928</td>\n",
       "      <td>5307394</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>PEX10</td>\n",
       "      <td>1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63236</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>None</td>\n",
       "      <td>KI270711.1</td>\n",
       "      <td>4612</td>\n",
       "      <td>29626</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63237</th>\n",
       "      <td>ENSG00000275987</td>\n",
       "      <td>U1</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>30437</td>\n",
       "      <td>30580</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63238</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>None</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>35407</td>\n",
       "      <td>35916</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63239</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>None</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>31698</td>\n",
       "      <td>32528</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63240</th>\n",
       "      <td>ENSG00000275405</td>\n",
       "      <td>U1</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63241 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id gene_name         chr    start      end  \\\n",
       "0      ENSG00000279928  DDX11L17           1   182696   184174   \n",
       "1      ENSG00000228037      None           1  2581560  2584533   \n",
       "2      ENSG00000142611    PRDM16           1  3069168  3438621   \n",
       "3      ENSG00000284616      None           1  5301928  5307394   \n",
       "4      ENSG00000157911     PEX10           1  2403964  2413797   \n",
       "...                ...       ...         ...      ...      ...   \n",
       "63236  ENSG00000271254      None  KI270711.1     4612    29626   \n",
       "63237  ENSG00000275987        U1  KI270713.1    30437    30580   \n",
       "63238  ENSG00000268674      None  KI270713.1    35407    35916   \n",
       "63239  ENSG00000277475      None  KI270713.1    31698    32528   \n",
       "63240  ENSG00000275405        U1  KI270713.1    21861    22024   \n",
       "\n",
       "                 gene_biotype  in_cgc  is_driver  \n",
       "0      unprocessed_pseudogene   False      False  \n",
       "1                      lncRNA   False      False  \n",
       "2              protein_coding    True      False  \n",
       "3                      lncRNA   False      False  \n",
       "4              protein_coding   False      False  \n",
       "...                       ...     ...        ...  \n",
       "63236          protein_coding   False      False  \n",
       "63237                   snRNA   False      False  \n",
       "63238          protein_coding   False      False  \n",
       "63239          protein_coding   False      False  \n",
       "63240                   snRNA   False      False  \n",
       "\n",
       "[63241 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirscribe-vcf-8VJm4tPg-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
