{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scripts.pipeline_steps import step5 as s5\n",
    "from scripts.notebook_operations import setup_notebook\n",
    "\n",
    "engine, g37, transcripts, genes, mirnas = setup_notebook()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_87132/4074695647.py:13: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df[['in_cgc', 'is_driver']] = df[['in_cgc', 'is_driver']].fillna(False).infer_objects(copy=False)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>wt_prediction</th>\n",
       "      <th>mut_prediction</th>\n",
       "      <th>pred_difference</th>\n",
       "      <th>vcf_id</th>\n",
       "      <th>mirna_accession</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mutation_context</th>\n",
       "      <th>mutsig</th>\n",
       "      <th>is_intron</th>\n",
       "      <th>is_gain</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>in_cgc</th>\n",
       "      <th>is_driver</th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>seed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0000265</td>\n",
       "      <td>0.411319</td>\n",
       "      <td>0.724802</td>\n",
       "      <td>0.313</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0000265</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-204-5p</td>\n",
       "      <td>miR-204-5p/211-5p</td>\n",
       "      <td>GGCATAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0003301</td>\n",
       "      <td>0.289311</td>\n",
       "      <td>0.662916</td>\n",
       "      <td>0.374</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0003301</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-33b-5p</td>\n",
       "      <td>miR-33-5p</td>\n",
       "      <td>CAATGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0004517</td>\n",
       "      <td>0.351466</td>\n",
       "      <td>0.623310</td>\n",
       "      <td>0.272</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0004517</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-106a-3p</td>\n",
       "      <td>miR-106a-3p</td>\n",
       "      <td>TAAGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0005882</td>\n",
       "      <td>0.550857</td>\n",
       "      <td>0.279106</td>\n",
       "      <td>-0.272</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0005882</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-548k</td>\n",
       "      <td>miR-548av-5p/548k/8054</td>\n",
       "      <td>GCAAAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD4005a_1_12844091_G_T_MIMAT0005933</td>\n",
       "      <td>0.650570</td>\n",
       "      <td>0.375474</td>\n",
       "      <td>-0.275</td>\n",
       "      <td>PD4005a</td>\n",
       "      <td>MIMAT0005933</td>\n",
       "      <td>not_found</td>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>SBS3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-1277-3p</td>\n",
       "      <td>miR-1277-3p</td>\n",
       "      <td>AAATACA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    id  wt_prediction  mut_prediction  \\\n",
       "0  PD4005a_1_12844091_G_T_MIMAT0000265       0.411319        0.724802   \n",
       "1  PD4005a_1_12844091_G_T_MIMAT0003301       0.289311        0.662916   \n",
       "2  PD4005a_1_12844091_G_T_MIMAT0004517       0.351466        0.623310   \n",
       "3  PD4005a_1_12844091_G_T_MIMAT0005882       0.550857        0.279106   \n",
       "4  PD4005a_1_12844091_G_T_MIMAT0005933       0.650570        0.375474   \n",
       "\n",
       "   pred_difference   vcf_id mirna_accession    gene_id mutation_context  \\\n",
       "0            0.313  PD4005a    MIMAT0000265  not_found          A[C>A]A   \n",
       "1            0.374  PD4005a    MIMAT0003301  not_found          A[C>A]A   \n",
       "2            0.272  PD4005a    MIMAT0004517  not_found          A[C>A]A   \n",
       "3           -0.272  PD4005a    MIMAT0005882  not_found          A[C>A]A   \n",
       "4           -0.275  PD4005a    MIMAT0005933  not_found          A[C>A]A   \n",
       "\n",
       "  mutsig  is_intron  is_gain  gene_name gene_biotype  in_cgc  is_driver  \\\n",
       "0   SBS3      False     True  not_found    not_found   False      False   \n",
       "1   SBS3      False     True  not_found    not_found   False      False   \n",
       "2   SBS3      False     True  not_found    not_found   False      False   \n",
       "3   SBS3      False    False  not_found    not_found   False      False   \n",
       "4   SBS3      False    False  not_found    not_found   False      False   \n",
       "\n",
       "        mirna_name            mirna_family     seed  \n",
       "0   hsa-miR-204-5p       miR-204-5p/211-5p  GGCATAG  \n",
       "1   hsa-miR-33b-5p               miR-33-5p  CAATGCA  \n",
       "2  hsa-miR-106a-3p             miR-106a-3p  TAAGAAG  \n",
       "3     hsa-miR-548k  miR-548av-5p/548k/8054  GCAAAAT  \n",
       "4  hsa-miR-1277-3p             miR-1277-3p  AAATACA  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_threshold, wt_threshold = (0.5 , 0.5)\n",
    "df = pd.read_parquet('data/predictions.parquet')\n",
    "\n",
    "\n",
    "\n",
    "df = s5.filter_rows_by_thresholds(df, mut_threshold, wt_threshold)\n",
    "\n",
    "df[\"is_gain\"] = df.mut_prediction > df.wt_prediction\n",
    "\n",
    "df = df.merge(genes[[\"gene_id\", \"gene_name\", \"gene_biotype\", \"in_cgc\", \"is_driver\"]], on=\"gene_id\", how=\"left\")\n",
    "df = df.merge(mirnas[[\"mirna_accession\", \"mirna_name\", \"mirna_family\",  \"seed\"]], on=\"mirna_accession\", how=\"left\")\n",
    "\n",
    "df[['in_cgc', 'is_driver']] = df[['in_cgc', 'is_driver']].fillna(False).infer_objects(copy=False)\n",
    "df[['gene_id', 'mutsig', \"gene_name\", \"gene_biotype\"]] = df[['gene_id', 'mutsig', \"gene_name\", \"gene_biotype\"]].fillna(\"not_found\").infer_objects(copy=False)\n",
    "\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genes' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mgenes\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'genes' is not defined"
     ]
    }
   ],
   "source": [
    "genes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirscribe-vcf-8VJm4tPg-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
