{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os \n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.stats import fisher_exact\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# read real and 10 synth data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# real = pd.read_parquet(\"data/synth/real.parquet\")\n",
    "# all = pd.read_parquet(\"data/synth/all.parquet\")\n",
    "\n",
    "# # sourcery skip: avoid-builtin-shadow\n",
    "# real_genes = set(real.gene_id)    \n",
    "# all_genes = set(all.gene_id)\n",
    "\n",
    "# # details\n",
    "# print(f\"Real genes: {len(real_genes)}\")\n",
    "# print(f\"All genes: {len(all_genes)}\")\n",
    "# print(f\"Real genes not in all: {len(real_genes - all_genes)}\")\n",
    "# print(f\"Crossection: {len(real_genes & all_genes)}\")\n",
    "\n",
    "# # taking crossection\n",
    "# common_genes = list(set(real.gene_id)  & set(all.gene_id))\n",
    "# real = real[real.gene_id.isin(common_genes)]\n",
    "# all = all[all.gene_id.isin(common_genes)]\n",
    "\n",
    "# # dropping not_found gene ids\n",
    "# real = real[real.gene_id != \"not_found\"]\n",
    "# all = all[all.gene_id != \"not_found\"]\n",
    "\n",
    "# real_2 = real.groupby(['gene_id', 'is_gene_upregulated']).size().unstack(fill_value=0)\n",
    "# synth_2 = all.groupby(['gene_id', 'is_gene_upregulated']).size().unstack(fill_value=0).div(10)\n",
    "# values = pd.merge(real_2, synth_2, on=\"gene_id\", suffixes=[\"_real\", \"_synth\"])\n",
    "# values.to_csv(\"data/synth/values.csv\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "values = pd.read_csv(\"data/synth/values.csv\")\n",
    "\n",
    "rename_dict = {\n",
    "    \"gene_id\": \"gene_id\",\n",
    "    \"False_real\": \"downregulated_real\",\n",
    "    \"True_real\": \"upregulated_real\",\n",
    "    \"False_synth\": \"downregulated_synth\",\n",
    "    \"True_synth\": \"upregulated_synth\"\n",
    "}\n",
    "\n",
    "values = values.rename(columns=rename_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_log2_odds_ratio(a, b, c, d, k=0.5):\n",
    "    # a, b, c, d are the four cells of the 2x2 contingency table\n",
    "    # k is the smoothing constant\n",
    "    #\n",
    "    odds_ratio = ((a + k) * (d + k)) / ((b + k) * (c + k))\n",
    "    return np.log2(odds_ratio)\n",
    "\n",
    "# laplace smoothing with k=0.5 (Jeffreys prior)\n",
    "values['log2_odds_ratio'] = values.apply(lambda row: calculate_log2_odds_ratio(\n",
    "    row['upregulated_real'], \n",
    "    row['downregulated_real'], \n",
    "    row['upregulated_synth'], \n",
    "    row['downregulated_synth']\n",
    "), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_fisher_test_vectorized(df, pseudocount=0.5):\n",
    "    # Add pseudocount to the table\n",
    "    table = np.array([\n",
    "        [df['upregulated_real'] + pseudocount, df['downregulated_real'] + pseudocount],\n",
    "        [df['upregulated_synth'] + pseudocount, df['downregulated_synth'] + pseudocount]\n",
    "    ]).transpose((2, 0, 1))  # reshape for 2x2 tables\n",
    "\n",
    "    p_values = np.zeros(len(df))\n",
    "\n",
    "    for i in range(len(df)):\n",
    "        _, p_values[i] = fisher_exact(table[i])\n",
    "\n",
    "    df['p_value'] = p_values\n",
    "    df['p_adj'] = multipletests(p_values, method='fdr_bh')[1]\n",
    "    \n",
    "    return df\n",
    "def add_z_score(df):\n",
    "    # Calculate mean and standard deviation of log2 odds ratios\n",
    "    mean_log2or = df['log2_odds_ratio'].mean()\n",
    "    std_log2or = df['log2_odds_ratio'].std()\n",
    "    \n",
    "    # Calculate Z-score\n",
    "    df['z_score'] = (df['log2_odds_ratio'] - mean_log2or) / std_log2or\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "df = perform_fisher_test_vectorized(values)\n",
    "df = add_z_score(df)\n",
    "\n",
    "\n",
    "log2_threshold = 0.32\n",
    "p_threshold = 0.05\n",
    "\n",
    "# df = df[(abs(df['log2_odds_ratio']) > log2_threshold) & (df['p_adj'] < p_threshold)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pyensembl import EnsemblRelease\n",
    "\n",
    "def get_gene_details(df, gene_id_column, release=75):\n",
    "    # Initialize Ensembl data for the specified release\n",
    "    ensembl = EnsemblRelease(release)\n",
    "    \n",
    "    def fetch_details(gene_id):\n",
    "        try:\n",
    "            gene = ensembl.gene_by_id(gene_id)\n",
    "            return {\n",
    "                'gene_id': gene.gene_id,\n",
    "                'gene_name': gene.gene_name,\n",
    "                'biotype': gene.biotype,\n",
    "                'chromosome': gene.contig,\n",
    "            }\n",
    "        except ValueError:\n",
    "            return {\n",
    "                'gene_id': None,\n",
    "                'gene_name': None,\n",
    "                'biotype': None,\n",
    "                'chromosome': None,\n",
    "            }\n",
    "    \n",
    "    # Apply the function to each gene ID in the column\n",
    "    gene_details = df[gene_id_column].apply(fetch_details)\n",
    "    \n",
    "    # Convert the result to a DataFrame\n",
    "    details_df = pd.DataFrame(gene_details.tolist())\n",
    "    \n",
    "\n",
    "    \n",
    "    return pd.merge(df, details_df, on='gene_id', how='left')\n",
    "result = get_gene_details(df, 'gene_id')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "result.drop(['upregulated_synth', 'downregulated_synth', 'upregulated_real', 'downregulated_real'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Load your data\n",
    "\n",
    "# Handle zero and very small p-values\n",
    "min_pvalue = result['p_adj'][result['p_adj'] > 0].min() / 2\n",
    "result['p_adj'] = result['p_adj'].replace(0, min_pvalue)\n",
    "\n",
    "# Calculate rank, handling potential infinite values\n",
    "result['rank'] = result['log2_odds_ratio'] * (-np.log10(result['p_adj']))\n",
    "result['rank'] = result['rank'].clip(lower=-1e10, upper=1e10)  # Clip extreme values\n",
    "\n",
    "# Add small random values to break ties\n",
    "result['rank'] = result['rank'] + np.random.random(len(result)) * 1e-10\n",
    "\n",
    "# Sort by rank\n",
    "df_sorted = result.sort_values('rank', ascending=False)\n",
    "\n",
    "# Create a Series with gene names as index and ranks as values\n",
    "ranked_genes = df_sorted.set_index('gene_name')['rank']\n",
    "\n",
    "# Run preranked GSEA\n",
    "pre_res = gp.prerank(rnk=ranked_genes, \n",
    "                     gene_sets='KEGG_2021_Human',\n",
    "                     threads=4,\n",
    "                     permutation_num=1000,\n",
    "                     min_size=10,\n",
    "                     max_size=1000,\n",
    "                     outdir='gsea_results/prerank_result',\n",
    "                     format='png')\n",
    "\n",
    "\n",
    "# Get the results\n",
    "results = pre_res.res2d\n",
    "sorted_results = results.sort_values('NES', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Term</th>\n",
       "      <th>ES</th>\n",
       "      <th>NES</th>\n",
       "      <th>NOM p-val</th>\n",
       "      <th>FDR q-val</th>\n",
       "      <th>FWER p-val</th>\n",
       "      <th>Tag %</th>\n",
       "      <th>Gene %</th>\n",
       "      <th>Lead_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>prerank</td>\n",
       "      <td>Glyoxylate and dicarboxylate metabolism</td>\n",
       "      <td>0.57868</td>\n",
       "      <td>1.651448</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.008535</td>\n",
       "      <td>0.006</td>\n",
       "      <td>2/13</td>\n",
       "      <td>2.18%</td>\n",
       "      <td>HAO2;ACAT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>prerank</td>\n",
       "      <td>Mannose type O-glycan biosynthesis</td>\n",
       "      <td>-0.270059</td>\n",
       "      <td>-0.454663</td>\n",
       "      <td>0.996929</td>\n",
       "      <td>0.999366</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2/11</td>\n",
       "      <td>29.79%</td>\n",
       "      <td>FKTN;FUT9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>prerank</td>\n",
       "      <td>Antigen processing and presentation</td>\n",
       "      <td>-0.347592</td>\n",
       "      <td>-0.623736</td>\n",
       "      <td>0.989</td>\n",
       "      <td>0.995351</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8/30</td>\n",
       "      <td>39.94%</td>\n",
       "      <td>HLA-DPA1;KIR3DL1;KLRD1;HLA-F;CD4;CANX;HSP90AA1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>prerank</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>-0.364853</td>\n",
       "      <td>-0.653153</td>\n",
       "      <td>0.977978</td>\n",
       "      <td>0.99539</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11/29</td>\n",
       "      <td>45.51%</td>\n",
       "      <td>PGK1;MINPP1;HK1;ADH1B;HKDC1;LDHB;ALDH3B2;ALDH9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>prerank</td>\n",
       "      <td>Ascorbate and aldarate metabolism</td>\n",
       "      <td>-0.383196</td>\n",
       "      <td>-0.658629</td>\n",
       "      <td>0.956566</td>\n",
       "      <td>0.997968</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8/14</td>\n",
       "      <td>40.46%</td>\n",
       "      <td>UGT2B7;UGT1A8;KL;RGN;UGT1A6;UGT2A3;ALDH9A1;UGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>prerank</td>\n",
       "      <td>Glutamatergic synapse</td>\n",
       "      <td>-0.697959</td>\n",
       "      <td>-1.302775</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.618783</td>\n",
       "      <td>0.927</td>\n",
       "      <td>36/83</td>\n",
       "      <td>18.82%</td>\n",
       "      <td>GNGT1;GNAQ;PLCB4;GRIA2;CACNA1C;GRIN2B;PRKCA;SL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>prerank</td>\n",
       "      <td>Taste transduction</td>\n",
       "      <td>-0.725356</td>\n",
       "      <td>-1.309327</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.697805</td>\n",
       "      <td>0.9</td>\n",
       "      <td>20/36</td>\n",
       "      <td>21.57%</td>\n",
       "      <td>PLCB4;CHRM3;PDE1C;CACNA1C;ASIC2;PDE1A;GABRA1;S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>prerank</td>\n",
       "      <td>African trypanosomiasis</td>\n",
       "      <td>-0.767753</td>\n",
       "      <td>-1.349436</td>\n",
       "      <td>0.014056</td>\n",
       "      <td>0.457877</td>\n",
       "      <td>0.709</td>\n",
       "      <td>3/19</td>\n",
       "      <td>1.09%</td>\n",
       "      <td>GNAQ;PLCB4;PRKCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>prerank</td>\n",
       "      <td>Folate biosynthesis</td>\n",
       "      <td>-0.859278</td>\n",
       "      <td>-1.444292</td>\n",
       "      <td>0.010215</td>\n",
       "      <td>0.0999</td>\n",
       "      <td>0.18</td>\n",
       "      <td>3/10</td>\n",
       "      <td>7.53%</td>\n",
       "      <td>GPHN;PAH;AKR1C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>prerank</td>\n",
       "      <td>Phototransduction</td>\n",
       "      <td>-0.847447</td>\n",
       "      <td>-1.461207</td>\n",
       "      <td>0.003027</td>\n",
       "      <td>0.136864</td>\n",
       "      <td>0.127</td>\n",
       "      <td>1/14</td>\n",
       "      <td>0.12%</td>\n",
       "      <td>GNGT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>299 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Name                                     Term        ES       NES  \\\n",
       "0    prerank  Glyoxylate and dicarboxylate metabolism   0.57868  1.651448   \n",
       "298  prerank       Mannose type O-glycan biosynthesis -0.270059 -0.454663   \n",
       "297  prerank      Antigen processing and presentation -0.347592 -0.623736   \n",
       "296  prerank             Glycolysis / Gluconeogenesis -0.364853 -0.653153   \n",
       "295  prerank        Ascorbate and aldarate metabolism -0.383196 -0.658629   \n",
       "..       ...                                      ...       ...       ...   \n",
       "5    prerank                    Glutamatergic synapse -0.697959 -1.302775   \n",
       "4    prerank                       Taste transduction -0.725356 -1.309327   \n",
       "3    prerank                  African trypanosomiasis -0.767753 -1.349436   \n",
       "2    prerank                      Folate biosynthesis -0.859278 -1.444292   \n",
       "1    prerank                        Phototransduction -0.847447 -1.461207   \n",
       "\n",
       "    NOM p-val FDR q-val FWER p-val  Tag %  Gene %  \\\n",
       "0    0.047619  0.008535      0.006   2/13   2.18%   \n",
       "298  0.996929  0.999366        1.0   2/11  29.79%   \n",
       "297     0.989  0.995351        1.0   8/30  39.94%   \n",
       "296  0.977978   0.99539        1.0  11/29  45.51%   \n",
       "295  0.956566  0.997968        1.0   8/14  40.46%   \n",
       "..        ...       ...        ...    ...     ...   \n",
       "5         0.0  0.618783      0.927  36/83  18.82%   \n",
       "4       0.007  0.697805        0.9  20/36  21.57%   \n",
       "3    0.014056  0.457877      0.709   3/19   1.09%   \n",
       "2    0.010215    0.0999       0.18   3/10   7.53%   \n",
       "1    0.003027  0.136864      0.127   1/14   0.12%   \n",
       "\n",
       "                                            Lead_genes  \n",
       "0                                           HAO2;ACAT1  \n",
       "298                                          FKTN;FUT9  \n",
       "297  HLA-DPA1;KIR3DL1;KLRD1;HLA-F;CD4;CANX;HSP90AA1...  \n",
       "296  PGK1;MINPP1;HK1;ADH1B;HKDC1;LDHB;ALDH3B2;ALDH9...  \n",
       "295  UGT2B7;UGT1A8;KL;RGN;UGT1A6;UGT2A3;ALDH9A1;UGT...  \n",
       "..                                                 ...  \n",
       "5    GNGT1;GNAQ;PLCB4;GRIA2;CACNA1C;GRIN2B;PRKCA;SL...  \n",
       "4    PLCB4;CHRM3;PDE1C;CACNA1C;ASIC2;PDE1A;GABRA1;S...  \n",
       "3                                     GNAQ;PLCB4;PRKCA  \n",
       "2                                      GPHN;PAH;AKR1C3  \n",
       "1                                                GNGT1  \n",
       "\n",
       "[299 rows x 10 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirscribe-vcf-zDi37XFo-py3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
