{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scripts.pipeline_steps import step5 as s5\n",
    "from scripts.notebook_operations import setup_notebook\n",
    "from scripts.globals import MUTSIG_PROBABILITIES\n",
    "\n",
    "engine, g37, transcripts, genes, mirnas = setup_notebook()\n",
    "\n",
    "\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "sns.set_style(\"whitegrid\")\n",
    "sns.set_palette(\"deep\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = s5.apply_step_5('results/test_simulated_1/results_test_simulated.csv', g37, MUTSIG_PROBABILITIES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1742"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_12594/278997352.py:21: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df = df.fillna(False)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>wt_prediction</th>\n",
       "      <th>mut_prediction</th>\n",
       "      <th>pred_difference</th>\n",
       "      <th>vcf_id</th>\n",
       "      <th>mirna_accession</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mutation_context</th>\n",
       "      <th>mutsig</th>\n",
       "      <th>is_intron</th>\n",
       "      <th>is_gain</th>\n",
       "      <th>is_gene_upregulated</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>is_driver_intogen</th>\n",
       "      <th>tier_cosmic</th>\n",
       "      <th>is_hallmark_cosmic</th>\n",
       "      <th>is_fusion</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tsupp</th>\n",
       "      <th>is_gene_of_interest</th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>seed</th>\n",
       "      <th>is_mirna_oncogene</th>\n",
       "      <th>is_mirna_tumor_suppressor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD3851a+_3_1992673_T_C_MIMAT0000075</td>\n",
       "      <td>0.429814</td>\n",
       "      <td>0.746308</td>\n",
       "      <td>0.316</td>\n",
       "      <td>PD3851a+</td>\n",
       "      <td>MIMAT0000075</td>\n",
       "      <td>not_found</td>\n",
       "      <td>G[T&gt;C]C</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-20a-5p</td>\n",
       "      <td>miR-17-5p/20-5p/93-5p/106-5p/519-3p</td>\n",
       "      <td>TACCTGC</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD3851a+_3_1992673_T_C_MIMAT0000093</td>\n",
       "      <td>0.412005</td>\n",
       "      <td>0.666354</td>\n",
       "      <td>0.254</td>\n",
       "      <td>PD3851a+</td>\n",
       "      <td>MIMAT0000093</td>\n",
       "      <td>not_found</td>\n",
       "      <td>G[T&gt;C]C</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-93-5p</td>\n",
       "      <td>miR-17-5p/20-5p/93-5p/106-5p/519-3p</td>\n",
       "      <td>TACCTGC</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD3851a+_3_1992673_T_C_MIMAT0000098</td>\n",
       "      <td>0.681526</td>\n",
       "      <td>0.446319</td>\n",
       "      <td>-0.235</td>\n",
       "      <td>PD3851a+</td>\n",
       "      <td>MIMAT0000098</td>\n",
       "      <td>not_found</td>\n",
       "      <td>G[T&gt;C]C</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-100-5p</td>\n",
       "      <td>miR-99-5p/100-5p</td>\n",
       "      <td>ACAAGTT</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD3851a+_3_1992673_T_C_MIMAT0000680</td>\n",
       "      <td>0.743227</td>\n",
       "      <td>0.279636</td>\n",
       "      <td>-0.464</td>\n",
       "      <td>PD3851a+</td>\n",
       "      <td>MIMAT0000680</td>\n",
       "      <td>not_found</td>\n",
       "      <td>G[T&gt;C]C</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-106b-5p</td>\n",
       "      <td>miR-17-5p/20-5p/93-5p/106-5p/519-3p</td>\n",
       "      <td>TCTGCAC</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD3851a+_3_1992673_T_C_MIMAT0001413</td>\n",
       "      <td>0.381469</td>\n",
       "      <td>0.706472</td>\n",
       "      <td>0.325</td>\n",
       "      <td>PD3851a+</td>\n",
       "      <td>MIMAT0001413</td>\n",
       "      <td>not_found</td>\n",
       "      <td>G[T&gt;C]C</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>not_found</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>hsa-miR-20b-5p</td>\n",
       "      <td>miR-17-5p/20-5p/93-5p/106-5p/519-3p</td>\n",
       "      <td>TACCTGC</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    id  wt_prediction  mut_prediction  \\\n",
       "0  PD3851a+_3_1992673_T_C_MIMAT0000075       0.429814        0.746308   \n",
       "1  PD3851a+_3_1992673_T_C_MIMAT0000093       0.412005        0.666354   \n",
       "2  PD3851a+_3_1992673_T_C_MIMAT0000098       0.681526        0.446319   \n",
       "3  PD3851a+_3_1992673_T_C_MIMAT0000680       0.743227        0.279636   \n",
       "4  PD3851a+_3_1992673_T_C_MIMAT0001413       0.381469        0.706472   \n",
       "\n",
       "   pred_difference    vcf_id mirna_accession    gene_id mutation_context  \\\n",
       "0            0.316  PD3851a+    MIMAT0000075  not_found          G[T>C]C   \n",
       "1            0.254  PD3851a+    MIMAT0000093  not_found          G[T>C]C   \n",
       "2           -0.235  PD3851a+    MIMAT0000098  not_found          G[T>C]C   \n",
       "3           -0.464  PD3851a+    MIMAT0000680  not_found          G[T>C]C   \n",
       "4            0.325  PD3851a+    MIMAT0001413  not_found          G[T>C]C   \n",
       "\n",
       "      mutsig  is_intron  is_gain  is_gene_upregulated  gene_name gene_biotype  \\\n",
       "0  not_found      False     True                False  not_found    not_found   \n",
       "1  not_found      False     True                False  not_found    not_found   \n",
       "2  not_found      False    False                 True  not_found    not_found   \n",
       "3  not_found      False    False                 True  not_found    not_found   \n",
       "4  not_found      False     True                False  not_found    not_found   \n",
       "\n",
       "  gene_description  is_driver_intogen tier_cosmic  is_hallmark_cosmic  \\\n",
       "0        not_found              False       False               False   \n",
       "1        not_found              False       False               False   \n",
       "2        not_found              False       False               False   \n",
       "3        not_found              False       False               False   \n",
       "4        not_found              False       False               False   \n",
       "\n",
       "   is_fusion  is_oncogene  is_tsupp  is_gene_of_interest       mirna_name  \\\n",
       "0      False        False     False                False   hsa-miR-20a-5p   \n",
       "1      False        False     False                False    hsa-miR-93-5p   \n",
       "2      False        False     False                False   hsa-miR-100-5p   \n",
       "3      False        False     False                False  hsa-miR-106b-5p   \n",
       "4      False        False     False                False   hsa-miR-20b-5p   \n",
       "\n",
       "                          mirna_family     seed  is_mirna_oncogene  \\\n",
       "0  miR-17-5p/20-5p/93-5p/106-5p/519-3p  TACCTGC               True   \n",
       "1  miR-17-5p/20-5p/93-5p/106-5p/519-3p  TACCTGC               True   \n",
       "2                     miR-99-5p/100-5p  ACAAGTT              False   \n",
       "3  miR-17-5p/20-5p/93-5p/106-5p/519-3p  TCTGCAC               True   \n",
       "4  miR-17-5p/20-5p/93-5p/106-5p/519-3p  TACCTGC              False   \n",
       "\n",
       "   is_mirna_tumor_suppressor  \n",
       "0                      False  \n",
       "1                      False  \n",
       "2                       True  \n",
       "3                      False  \n",
       "4                      False  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "mut_threshold, wt_threshold = (0.5 , 0.5)\n",
    "df = s5.filter_rows_by_thresholds(df, mut_threshold, wt_threshold)\n",
    "\n",
    "# add cols\n",
    "df[\"is_gain\"] = df.mut_prediction > df.wt_prediction\n",
    "df[\"is_gene_upregulated\"] = ~df.is_gain\n",
    "\n",
    "\n",
    "\n",
    "# merge \n",
    "df = df.merge(genes, on=\"gene_id\", how=\"left\")\n",
    "df = df.merge(mirnas, on=\"mirna_accession\", how=\"left\")\n",
    "df.drop(columns=[\"chr\", \"start\", \"end\", \"sequence\", \"conservation\", \"mirna_name_cmc\"], inplace=True)\n",
    "\n",
    "# fill NAs\n",
    "df[\"mutsig\"] = df[\"mutsig\"].fillna(\"not_found\")\n",
    "df[\"gene_id\"] = df[\"gene_id\"].fillna(\"not_found\")\n",
    "df[\"gene_name\"] = df[\"gene_name\"].fillna(\"not_found\")\n",
    "df[\"gene_biotype\"] = df[\"gene_biotype\"].fillna(\"not_found\")\n",
    "df[\"gene_description\"] = df[\"gene_description\"].fillna(\"not_found\")\n",
    "df = df.fillna(False)\n",
    "\n",
    "df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirscribe-vcf-8VJm4tPg-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
