{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/nazif/thesis/mirscribe-vcf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nazif/.cache/pypoetry/virtualenvs/mirscribe-vcf-8VJm4tPg-py3.10/lib/python3.10/site-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "cd ../.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "from sqlalchemy import create_engine\n",
    "pyensembl = create_engine('sqlite:////home/nazif/.cache/pyensembl/GRCh38/ensembl111/Homo_sapiens.GRCh38.111.gtf.db')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_33248/2886532373.py:19: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  gsea.fillna(False, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>is_tnfa_signaling_via_nfkb</th>\n",
       "      <th>is_hypoxia</th>\n",
       "      <th>is_cholesterol_homeostasis</th>\n",
       "      <th>is_mitotic_spindle</th>\n",
       "      <th>is_wnt_beta_catenin_signaling</th>\n",
       "      <th>is_tgf_beta_signaling</th>\n",
       "      <th>is_il6_jak_stat3_signaling</th>\n",
       "      <th>is_dna_repair</th>\n",
       "      <th>is_g2m_checkpoint</th>\n",
       "      <th>...</th>\n",
       "      <th>is_heme_metabolism</th>\n",
       "      <th>is_coagulation</th>\n",
       "      <th>is_il2_stat5_signaling</th>\n",
       "      <th>is_bile_acid_metabolism</th>\n",
       "      <th>is_peroxisome</th>\n",
       "      <th>is_allograft_rejection</th>\n",
       "      <th>is_spermatogenesis</th>\n",
       "      <th>is_kras_signaling_up</th>\n",
       "      <th>is_kras_signaling_dn</th>\n",
       "      <th>is_pancreas_beta_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AARS1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABAT</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4379</th>\n",
       "      <td>ZNRF4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4380</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4381</th>\n",
       "      <td>ZW10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4382</th>\n",
       "      <td>ZWINT</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4383</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4384 rows × 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  is_tnfa_signaling_via_nfkb  is_hypoxia  \\\n",
       "0          A2M                       False       False   \n",
       "1         AAAS                       False       False   \n",
       "2        AADAT                       False       False   \n",
       "3        AARS1                       False       False   \n",
       "4         ABAT                       False       False   \n",
       "...        ...                         ...         ...   \n",
       "4379     ZNRF4                       False       False   \n",
       "4380      ZPBP                       False       False   \n",
       "4381      ZW10                       False       False   \n",
       "4382     ZWINT                       False       False   \n",
       "4383       ZYX                       False       False   \n",
       "\n",
       "      is_cholesterol_homeostasis  is_mitotic_spindle  \\\n",
       "0                          False               False   \n",
       "1                          False               False   \n",
       "2                          False               False   \n",
       "3                          False               False   \n",
       "4                          False               False   \n",
       "...                          ...                 ...   \n",
       "4379                       False               False   \n",
       "4380                       False               False   \n",
       "4381                       False               False   \n",
       "4382                       False               False   \n",
       "4383                       False               False   \n",
       "\n",
       "      is_wnt_beta_catenin_signaling  is_tgf_beta_signaling  \\\n",
       "0                             False                  False   \n",
       "1                             False                  False   \n",
       "2                             False                  False   \n",
       "3                             False                  False   \n",
       "4                             False                  False   \n",
       "...                             ...                    ...   \n",
       "4379                          False                  False   \n",
       "4380                          False                  False   \n",
       "4381                          False                  False   \n",
       "4382                          False                  False   \n",
       "4383                          False                  False   \n",
       "\n",
       "      is_il6_jak_stat3_signaling  is_dna_repair  is_g2m_checkpoint  ...  \\\n",
       "0                           True          False              False  ...   \n",
       "1                          False           True              False  ...   \n",
       "2                          False          False              False  ...   \n",
       "3                          False          False              False  ...   \n",
       "4                          False          False              False  ...   \n",
       "...                          ...            ...                ...  ...   \n",
       "4379                       False          False              False  ...   \n",
       "4380                       False          False              False  ...   \n",
       "4381                       False          False              False  ...   \n",
       "4382                       False           True              False  ...   \n",
       "4383                       False          False              False  ...   \n",
       "\n",
       "      is_heme_metabolism  is_coagulation  is_il2_stat5_signaling  \\\n",
       "0                  False            True                   False   \n",
       "1                  False           False                   False   \n",
       "2                  False           False                   False   \n",
       "3                  False           False                   False   \n",
       "4                  False           False                   False   \n",
       "...                  ...             ...                     ...   \n",
       "4379               False           False                   False   \n",
       "4380               False           False                   False   \n",
       "4381               False           False                   False   \n",
       "4382               False           False                   False   \n",
       "4383               False           False                   False   \n",
       "\n",
       "      is_bile_acid_metabolism  is_peroxisome  is_allograft_rejection  \\\n",
       "0                       False          False                   False   \n",
       "1                       False          False                   False   \n",
       "2                       False          False                   False   \n",
       "3                       False          False                    True   \n",
       "4                       False          False                   False   \n",
       "...                       ...            ...                     ...   \n",
       "4379                    False          False                   False   \n",
       "4380                    False          False                   False   \n",
       "4381                    False          False                   False   \n",
       "4382                    False          False                   False   \n",
       "4383                    False          False                   False   \n",
       "\n",
       "      is_spermatogenesis  is_kras_signaling_up  is_kras_signaling_dn  \\\n",
       "0                  False                 False                 False   \n",
       "1                  False                 False                 False   \n",
       "2                  False                 False                 False   \n",
       "3                  False                 False                 False   \n",
       "4                  False                 False                 False   \n",
       "...                  ...                   ...                   ...   \n",
       "4379                True                 False                 False   \n",
       "4380                True                 False                 False   \n",
       "4381               False                 False                 False   \n",
       "4382               False                 False                 False   \n",
       "4383               False                 False                 False   \n",
       "\n",
       "      is_pancreas_beta_cells  \n",
       "0                      False  \n",
       "1                      False  \n",
       "2                      False  \n",
       "3                      False  \n",
       "4                      False  \n",
       "...                      ...  \n",
       "4379                   False  \n",
       "4380                   False  \n",
       "4381                   False  \n",
       "4382                   False  \n",
       "4383                   False  \n",
       "\n",
       "[4384 rows x 51 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "with open(\"data/gsea/h.all.v2023.2.Hs.json\") as file:\n",
    "    gsea_data = json.load(file)\n",
    "\n",
    "# Create a dictionary to store the data\n",
    "data_dict = {}\n",
    "\n",
    "# Iterate over each hallmark property\n",
    "for hallmark, hallmark_data in gsea_data.items():\n",
    "    # Extract the gene symbols for the current hallmark\n",
    "    gene_symbols = hallmark_data['geneSymbols']\n",
    "    \n",
    "    # Add the gene symbols as a column in the data dictionary\n",
    "    data_dict[hallmark] = pd.Series(True, index=gene_symbols)\n",
    "\n",
    "# Create the DataFrame from the data dictionary\n",
    "gsea = pd.DataFrame(data_dict)\n",
    "\n",
    "# Fill missing values with False\n",
    "gsea.fillna(False, inplace=True)\n",
    "\n",
    "# Reset the index to make gene IDs as rows\n",
    "gsea.reset_index(inplace=True)\n",
    "gsea.rename(columns={'index': 'gene_name'}, inplace=True)\n",
    "\n",
    "def clean_column_name(col_name):\n",
    "    if col_name.startswith(\"HALLMARK_\"):\n",
    "        return \"is_\" + col_name[len(\"HALLMARK_\"):].lower()\n",
    "    return col_name.lower()\n",
    "\n",
    "# Apply the function to all column names\n",
    "gsea.rename(columns=lambda col: clean_column_name(col), inplace=True)\n",
    "\n",
    "gsea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.read_sql(\"gene\", pyensembl, columns=[\"gene_id\", \"gene_name\"])\n",
    "df = gsea.merge(genes, on=\"gene_name\", how=\"left\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>is_tnfa_signaling_via_nfkb</th>\n",
       "      <th>is_hypoxia</th>\n",
       "      <th>is_cholesterol_homeostasis</th>\n",
       "      <th>is_mitotic_spindle</th>\n",
       "      <th>is_wnt_beta_catenin_signaling</th>\n",
       "      <th>is_tgf_beta_signaling</th>\n",
       "      <th>is_il6_jak_stat3_signaling</th>\n",
       "      <th>is_dna_repair</th>\n",
       "      <th>is_g2m_checkpoint</th>\n",
       "      <th>...</th>\n",
       "      <th>is_coagulation</th>\n",
       "      <th>is_il2_stat5_signaling</th>\n",
       "      <th>is_bile_acid_metabolism</th>\n",
       "      <th>is_peroxisome</th>\n",
       "      <th>is_allograft_rejection</th>\n",
       "      <th>is_spermatogenesis</th>\n",
       "      <th>is_kras_signaling_up</th>\n",
       "      <th>is_kras_signaling_dn</th>\n",
       "      <th>is_pancreas_beta_cells</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2386</th>\n",
       "      <td>METTL7B</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4008</th>\n",
       "      <td>THEG</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  is_tnfa_signaling_via_nfkb  is_hypoxia  \\\n",
       "2386   METTL7B                       False       False   \n",
       "4008      THEG                       False       False   \n",
       "\n",
       "      is_cholesterol_homeostasis  is_mitotic_spindle  \\\n",
       "2386                       False               False   \n",
       "4008                       False               False   \n",
       "\n",
       "      is_wnt_beta_catenin_signaling  is_tgf_beta_signaling  \\\n",
       "2386                          False                  False   \n",
       "4008                          False                  False   \n",
       "\n",
       "      is_il6_jak_stat3_signaling  is_dna_repair  is_g2m_checkpoint  ...  \\\n",
       "2386                       False          False              False  ...   \n",
       "4008                       False          False              False  ...   \n",
       "\n",
       "      is_coagulation  is_il2_stat5_signaling  is_bile_acid_metabolism  \\\n",
       "2386           False                   False                    False   \n",
       "4008           False                   False                    False   \n",
       "\n",
       "      is_peroxisome  is_allograft_rejection  is_spermatogenesis  \\\n",
       "2386          False                   False               False   \n",
       "4008          False                   False                True   \n",
       "\n",
       "      is_kras_signaling_up  is_kras_signaling_dn  is_pancreas_beta_cells  \\\n",
       "2386                 False                 False                   False   \n",
       "4008                 False                 False                   False   \n",
       "\n",
       "      gene_id  \n",
       "2386      NaN  \n",
       "4008      NaN  \n",
       "\n",
       "[2 rows x 52 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.gene_id.isna()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manually taken from ENSEMBL\n",
    "df.loc[df.gene_name == 'METTL7B', 'gene_id'] = 'ENSG00000163575'\n",
    "df.loc[df.gene_name == 'THEG', 'gene_id'] = 'ENSG00000105549'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name                               0\n",
       "is_tnfa_signaling_via_nfkb              0\n",
       "is_hypoxia                              0\n",
       "is_cholesterol_homeostasis              0\n",
       "is_mitotic_spindle                      0\n",
       "is_wnt_beta_catenin_signaling           0\n",
       "is_tgf_beta_signaling                   0\n",
       "is_il6_jak_stat3_signaling              0\n",
       "is_dna_repair                           0\n",
       "is_g2m_checkpoint                       0\n",
       "is_apoptosis                            0\n",
       "is_notch_signaling                      0\n",
       "is_adipogenesis                         0\n",
       "is_estrogen_response_early              0\n",
       "is_estrogen_response_late               0\n",
       "is_androgen_response                    0\n",
       "is_myogenesis                           0\n",
       "is_protein_secretion                    0\n",
       "is_interferon_alpha_response            0\n",
       "is_interferon_gamma_response            0\n",
       "is_apical_junction                      0\n",
       "is_apical_surface                       0\n",
       "is_hedgehog_signaling                   0\n",
       "is_complement                           0\n",
       "is_unfolded_protein_response            0\n",
       "is_pi3k_akt_mtor_signaling              0\n",
       "is_mtorc1_signaling                     0\n",
       "is_e2f_targets                          0\n",
       "is_myc_targets_v1                       0\n",
       "is_myc_targets_v2                       0\n",
       "is_epithelial_mesenchymal_transition    0\n",
       "is_inflammatory_response                0\n",
       "is_xenobiotic_metabolism                0\n",
       "is_fatty_acid_metabolism                0\n",
       "is_oxidative_phosphorylation            0\n",
       "is_glycolysis                           0\n",
       "is_reactive_oxygen_species_pathway      0\n",
       "is_p53_pathway                          0\n",
       "is_uv_response_up                       0\n",
       "is_uv_response_dn                       0\n",
       "is_angiogenesis                         0\n",
       "is_heme_metabolism                      0\n",
       "is_coagulation                          0\n",
       "is_il2_stat5_signaling                  0\n",
       "is_bile_acid_metabolism                 0\n",
       "is_peroxisome                           0\n",
       "is_allograft_rejection                  0\n",
       "is_spermatogenesis                      0\n",
       "is_kras_signaling_up                    0\n",
       "is_kras_signaling_dn                    0\n",
       "is_pancreas_beta_cells                  0\n",
       "gene_id                                 0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"data/gsea/gsea.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirscribe-vcf-8VJm4tPg-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
