{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/nazif/thesis/mirscribe-vcf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nazif/.cache/pypoetry/virtualenvs/mirscribe-vcf-8VJm4tPg-py3.10/lib/python3.10/site-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "cd ../.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>entrez_gene_id</th>\n",
       "      <th>grch37_isoform</th>\n",
       "      <th>grch37_refseq</th>\n",
       "      <th>grch38_isoform</th>\n",
       "      <th>grch38_refseq</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>occurrence_within_resources</th>\n",
       "      <th>oncokb_annotated</th>\n",
       "      <th>msk_impact</th>\n",
       "      <th>msk_heme</th>\n",
       "      <th>foundation_one</th>\n",
       "      <th>foundation_one_heme</th>\n",
       "      <th>vogelstein</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "      <th>gene_aliases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>25</td>\n",
       "      <td>ENST00000318560</td>\n",
       "      <td>NM_005157.4</td>\n",
       "      <td>ENST00000318560</td>\n",
       "      <td>NM_005157.4</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ABL, JTK7, c-ABL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>207</td>\n",
       "      <td>ENST00000349310</td>\n",
       "      <td>NM_001014431.1</td>\n",
       "      <td>ENST00000349310</td>\n",
       "      <td>NM_001014431.1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>AKT, PKB, PRKBA, RAC, RAC-alpha</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALK</td>\n",
       "      <td>238</td>\n",
       "      <td>ENST00000389048</td>\n",
       "      <td>NM_004304.4</td>\n",
       "      <td>ENST00000389048</td>\n",
       "      <td>NM_004304.4</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>CD246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMER1</td>\n",
       "      <td>139285</td>\n",
       "      <td>ENST00000330258</td>\n",
       "      <td>NM_152424.3</td>\n",
       "      <td>ENST00000374869</td>\n",
       "      <td>NM_152424.3</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>FAM123B, FLJ39827, RP11-403E24.2, WTX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APC</td>\n",
       "      <td>324</td>\n",
       "      <td>ENST00000257430</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>ENST00000257430</td>\n",
       "      <td>NM_000038.5</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>DP2.5, PPP1R46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol  entrez_gene_id   grch37_isoform   grch37_refseq  \\\n",
       "0        ABL1              25  ENST00000318560     NM_005157.4   \n",
       "1        AKT1             207  ENST00000349310  NM_001014431.1   \n",
       "2         ALK             238  ENST00000389048     NM_004304.4   \n",
       "3       AMER1          139285  ENST00000330258     NM_152424.3   \n",
       "4         APC             324  ENST00000257430     NM_000038.5   \n",
       "\n",
       "    grch38_isoform   grch38_refseq is_oncogene is_tumor_suppressor  \\\n",
       "0  ENST00000318560     NM_005157.4         Yes                  No   \n",
       "1  ENST00000349310  NM_001014431.1         Yes                  No   \n",
       "2  ENST00000389048     NM_004304.4         Yes                  No   \n",
       "3  ENST00000374869     NM_152424.3          No                 Yes   \n",
       "4  ENST00000257430     NM_000038.5          No                 Yes   \n",
       "\n",
       "   occurrence_within_resources oncokb_annotated msk_impact msk_heme  \\\n",
       "0                            7              Yes        Yes      Yes   \n",
       "1                            7              Yes        Yes      Yes   \n",
       "2                            7              Yes        Yes      Yes   \n",
       "3                            7              Yes        Yes      Yes   \n",
       "4                            7              Yes        Yes      Yes   \n",
       "\n",
       "  foundation_one foundation_one_heme vogelstein is_cosmic_tier_1  \\\n",
       "0            Yes                 Yes        Yes              Yes   \n",
       "1            Yes                 Yes        Yes              Yes   \n",
       "2            Yes                 Yes        Yes              Yes   \n",
       "3            Yes                 Yes        Yes              Yes   \n",
       "4            Yes                 Yes        Yes              Yes   \n",
       "\n",
       "                            gene_aliases  \n",
       "0                       ABL, JTK7, c-ABL  \n",
       "1        AKT, PKB, PRKBA, RAC, RAC-alpha  \n",
       "2                                  CD246  \n",
       "3  FAM123B, FLJ39827, RP11-403E24.2, WTX  \n",
       "4                         DP2.5, PPP1R46  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kb = pd.read_csv(\"data/oncokb/cancerGeneList.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Create a dictionary mapping the old column names to the new ones\n",
    "new_column_names = {\n",
    "    'Hugo Symbol': 'gene_symbol',\n",
    "    'Entrez Gene ID': 'entrez_gene_id',\n",
    "    'GRCh37 Isoform': 'grch37_isoform',\n",
    "    'GRCh37 RefSeq': 'grch37_refseq',\n",
    "    'GRCh38 Isoform': 'grch38_isoform',\n",
    "    'GRCh38 RefSeq': 'grch38_refseq',\n",
    "    'Is Oncogene': 'is_oncogene',\n",
    "    'Is Tumor Suppressor Gene': 'is_tumor_suppressor',\n",
    "    '# of occurrence within resources (Column J-P)': 'occurrence_within_resources',\n",
    "    'OncoKB Annotated': 'oncokb_annotated',\n",
    "    'MSK-IMPACT': 'msk_impact',\n",
    "    'MSK-HEME': 'msk_heme',\n",
    "    'FOUNDATION ONE': 'foundation_one',\n",
    "    'FOUNDATION ONE HEME': 'foundation_one_heme',\n",
    "    'Vogelstein': 'vogelstein',\n",
    "    'COSMIC CGC (v99)': 'is_cosmic_tier_1',\n",
    "    'Gene Aliases': 'gene_aliases'\n",
    "}\n",
    "\n",
    "# Rename the columns using the dictionary\n",
    "kb = kb.rename(columns=new_column_names)\n",
    "\n",
    "kb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = (pd.read_csv(\"data/oncokb/entrez_id_to_ensg.tsv\", sep=\"\\t\")\n",
    "       .dropna(how='all')\n",
    "       .rename(columns={'NCBI Gene ID': 'ncbi_gene_id', 'Ensembl gene ID': 'ensembl_gene_id'})\n",
    ")\n",
    "\n",
    "ids.ncbi_gene_id = ids.fillna(0).ncbi_gene_id.astype(int)\n",
    "\n",
    "kb = kb.merge(ids, left_on=\"entrez_gene_id\", right_on=\"ncbi_gene_id\", how=\"left\")\n",
    "kb.drop(\"ncbi_gene_id\", axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "826    6955\n",
       "827    6957\n",
       "828    6964\n",
       "829    6965\n",
       "Name: entrez_gene_id, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_genes = kb[((kb[\"is_oncogene\"] == \"Yes\") | (kb[\"is_tumor_suppressor\"] == \"Yes\")) & (kb.ensembl_gene_id.isna())].entrez_gene_id\n",
    "missing_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# checking hgnc if these genes are available"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>MANE Select Ensembl transcript ID (supplied by NCBI)</th>\n",
       "      <th>Ensembl gene ID</th>\n",
       "      <th>Gene group name</th>\n",
       "      <th>NCBI Gene ID</th>\n",
       "      <th>Locus type</th>\n",
       "      <th>Enzyme IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39737</th>\n",
       "      <td>TRA</td>\n",
       "      <td>T cell receptor alpha locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T cell receptor alpha locus at 14q11.2</td>\n",
       "      <td>6955.0</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39963</th>\n",
       "      <td>TRB</td>\n",
       "      <td>T cell receptor beta locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T cell receptor beta locus at 7q34</td>\n",
       "      <td>6957.0</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40091</th>\n",
       "      <td>TRD</td>\n",
       "      <td>T cell receptor delta locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T cell receptor delta locus at 14q11.2</td>\n",
       "      <td>6964.0</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40201</th>\n",
       "      <td>TRG</td>\n",
       "      <td>T cell receptor gamma locus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T cell receptor gamma locus at 7p14</td>\n",
       "      <td>6965.0</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Approved symbol                Approved name  \\\n",
       "39737             TRA  T cell receptor alpha locus   \n",
       "39963             TRB   T cell receptor beta locus   \n",
       "40091             TRD  T cell receptor delta locus   \n",
       "40201             TRG  T cell receptor gamma locus   \n",
       "\n",
       "      MANE Select Ensembl transcript ID (supplied by NCBI) Ensembl gene ID  \\\n",
       "39737                                                NaN               NaN   \n",
       "39963                                                NaN               NaN   \n",
       "40091                                                NaN               NaN   \n",
       "40201                                                NaN               NaN   \n",
       "\n",
       "                              Gene group name  NCBI Gene ID  \\\n",
       "39737  T cell receptor alpha locus at 14q11.2        6955.0   \n",
       "39963      T cell receptor beta locus at 7q34        6957.0   \n",
       "40091  T cell receptor delta locus at 14q11.2        6964.0   \n",
       "40201     T cell receptor gamma locus at 7p14        6965.0   \n",
       "\n",
       "                      Locus type Enzyme IDs  \n",
       "39737  gene with protein product        NaN  \n",
       "39963  gene with protein product        NaN  \n",
       "40091  gene with protein product        NaN  \n",
       "40201  gene with protein product        NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgnc = pd.read_csv(\"data/hgnc/hgnc.tsv\", sep=\"\\t\")\n",
    "hgnc[hgnc[\"NCBI Gene ID\"].isin(missing_genes)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# these genes have no reference in both GRCh37 & 38."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop the missing genes\n",
    "kb.drop(kb[((kb[\"is_oncogene\"] == \"Yes\") | (kb[\"is_tumor_suppressor\"] == \"Yes\")) & (kb.ensembl_gene_id.isna())].index, inplace=True)\n",
    "\n",
    "# filter cols\n",
    "cols_to_keep = ['gene_symbol', 'ensembl_gene_id', 'is_oncogene', 'is_tumor_suppressor', 'is_cosmic_tier_1']\n",
    "kb = kb[cols_to_keep]\n",
    "\n",
    "# change column types to bool\n",
    "for column in kb.columns:\n",
    "    if column.startswith(\"is_\"):\n",
    "        kb[column] = kb[column].map({\"Yes\": True, \"No\": False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>ENSG00000142208</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALK</td>\n",
       "      <td>ENSG00000171094</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMER1</td>\n",
       "      <td>ENSG00000184675</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APC</td>\n",
       "      <td>ENSG00000134982</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol  ensembl_gene_id  is_oncogene  is_tumor_suppressor  \\\n",
       "0        ABL1  ENSG00000097007         True                False   \n",
       "1        AKT1  ENSG00000142208         True                False   \n",
       "2         ALK  ENSG00000171094         True                False   \n",
       "3       AMER1  ENSG00000184675        False                 True   \n",
       "4         APC  ENSG00000134982        False                 True   \n",
       "\n",
       "   is_cosmic_tier_1  \n",
       "0              True  \n",
       "1              True  \n",
       "2              True  \n",
       "3              True  \n",
       "4              True  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "kb.to_csv(\"data/oncokb/oncokb.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirscribe-vcf-8VJm4tPg-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
