{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/nazif/thesis/mirscribe-vcf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nazif/.cache/pypoetry/virtualenvs/mirscribe-vcf-8VJm4tPg-py3.10/lib/python3.10/site-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "cd ../.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import create_engine, inspect\n",
    "from sqlalchemy import create_engine, Column, Integer, String, ForeignKey, Float, UniqueConstraint\n",
    "from sqlalchemy.ext.declarative import declarative_base\n",
    "from sqlalchemy.orm import relationship\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# db engine\n",
    "engine = create_engine('sqlite:///data/db/mirscribe.db')\n",
    "\n",
    "# pyensembl db engine\n",
    "pyensembl = create_engine('sqlite:////home/nazif/.cache/pyensembl/GRCh38/ensembl111/Homo_sapiens.GRCh38.111.gtf.db')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# mirnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>mirna_accession</th>\n",
       "      <th>sequence</th>\n",
       "      <th>mirna_family</th>\n",
       "      <th>conservation</th>\n",
       "      <th>seed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>MIMAT0000062</td>\n",
       "      <td>AACTATACAACCTACTACCTCA</td>\n",
       "      <td>let-7-5p/98-5p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ACTATAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>MIMAT0000063</td>\n",
       "      <td>AACCACACAACCTACTACCTCA</td>\n",
       "      <td>let-7-5p/98-5p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ACCACAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7c-5p</td>\n",
       "      <td>MIMAT0000064</td>\n",
       "      <td>AACCATACAACCTACTACCTCA</td>\n",
       "      <td>let-7-5p/98-5p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ACCATAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7d-5p</td>\n",
       "      <td>MIMAT0000065</td>\n",
       "      <td>AACTATGCAACCTACTACCTCT</td>\n",
       "      <td>let-7-5p/98-5p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ACTATGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7e-5p</td>\n",
       "      <td>MIMAT0000066</td>\n",
       "      <td>AACTATACAACCTCCTACCTCA</td>\n",
       "      <td>let-7-5p/98-5p</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ACTATAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      mirna_name mirna_accession                sequence    mirna_family  \\\n",
       "0  hsa-let-7a-5p    MIMAT0000062  AACTATACAACCTACTACCTCA  let-7-5p/98-5p   \n",
       "1  hsa-let-7b-5p    MIMAT0000063  AACCACACAACCTACTACCTCA  let-7-5p/98-5p   \n",
       "2  hsa-let-7c-5p    MIMAT0000064  AACCATACAACCTACTACCTCA  let-7-5p/98-5p   \n",
       "3  hsa-let-7d-5p    MIMAT0000065  AACTATGCAACCTACTACCTCT  let-7-5p/98-5p   \n",
       "4  hsa-let-7e-5p    MIMAT0000066  AACTATACAACCTCCTACCTCA  let-7-5p/98-5p   \n",
       "\n",
       "   conservation     seed  \n",
       "0           2.0  ACTATAC  \n",
       "1           2.0  ACCACAC  \n",
       "2           2.0  ACCATAC  \n",
       "3           2.0  ACTATGC  \n",
       "4           2.0  ACTATAC  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_df = pd.read_csv('data/mirna/mirna.csv')\n",
    "mirna_df[\"seed\"] = mirna_df.sequence.str[1:8]\n",
    "\n",
    "mirna_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mirna_name         0\n",
       "mirna_accession    0\n",
       "sequence           0\n",
       "mirna_family       0\n",
       "conservation       0\n",
       "seed               0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_df.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>1</td>\n",
       "      <td>182696</td>\n",
       "      <td>184174</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000228037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>2581560</td>\n",
       "      <td>2584533</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000284616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5301928</td>\n",
       "      <td>5307394</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>PEX10</td>\n",
       "      <td>1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name chr    start      end            gene_biotype\n",
       "0  ENSG00000279928  DDX11L17   1   182696   184174  unprocessed_pseudogene\n",
       "1  ENSG00000228037       NaN   1  2581560  2584533                  lncRNA\n",
       "2  ENSG00000142611    PRDM16   1  3069168  3438621          protein_coding\n",
       "3  ENSG00000284616       NaN   1  5301928  5307394                  lncRNA\n",
       "4  ENSG00000157911     PEX10   1  2403964  2413797          protein_coding"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_sql(\"gene\", pyensembl, columns=[\"gene_id\", \"gene_name\", \"seqname\", \"start\", \"end\", \"gene_biotype\"])\n",
    "\n",
    "genes = genes.replace(r'^\\s*$', np.nan, regex=True)\n",
    "# /tmp/ipykernel_91379/1564060209.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
    "genes = genes.infer_objects(copy=False)\n",
    "\n",
    "genes.rename(columns={\"seqname\": \"chr\"}, inplace=True)\n",
    "\n",
    "genes.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431</td>\n",
       "      <td>DDX11L17-201</td>\n",
       "      <td>1</td>\n",
       "      <td>182696</td>\n",
       "      <td>184174</td>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000424215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>2581560</td>\n",
       "      <td>2584533</td>\n",
       "      <td>ENSG00000228037</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000511072</td>\n",
       "      <td>PRDM16-206</td>\n",
       "      <td>1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3434342</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000607632</td>\n",
       "      <td>PRDM16-210</td>\n",
       "      <td>1</td>\n",
       "      <td>3069183</td>\n",
       "      <td>3186591</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>retained_intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000378391</td>\n",
       "      <td>PRDM16-203</td>\n",
       "      <td>1</td>\n",
       "      <td>3069197</td>\n",
       "      <td>3435421</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id transcript_name chr    start      end          gene_id  \\\n",
       "0  ENST00000624431    DDX11L17-201   1   182696   184174  ENSG00000279928   \n",
       "1  ENST00000424215             NaN   1  2581560  2584533  ENSG00000228037   \n",
       "2  ENST00000511072      PRDM16-206   1  3069168  3434342  ENSG00000142611   \n",
       "3  ENST00000607632      PRDM16-210   1  3069183  3186591  ENSG00000142611   \n",
       "4  ENST00000378391      PRDM16-203   1  3069197  3435421  ENSG00000142611   \n",
       "\n",
       "       transcript_biotype  \n",
       "0  unprocessed_pseudogene  \n",
       "1                  lncRNA  \n",
       "2          protein_coding  \n",
       "3         retained_intron  \n",
       "4          protein_coding  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts = pd.read_sql(\"transcript\", pyensembl, columns=[\"transcript_id\", \"transcript_name\", \"seqname\", \"start\", \"end\", \"gene_id\", \"transcript_biotype\"])\n",
    "\n",
    "transcripts = transcripts.replace(r'^\\s*$', np.nan, regex=True)\n",
    "\n",
    "transcripts.rename(columns={\"seqname\": \"chr\"}, inplace=True)\n",
    "transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "transcript_id             0\n",
       "transcript_name       31617\n",
       "chr                       0\n",
       "start                     0\n",
       "end                       0\n",
       "gene_id                   0\n",
       "transcript_biotype        0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# gsea hallmark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_14504/3987882754.py:23: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  gsea.fillna(False, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>tnfa_signaling_via_nfkb</th>\n",
       "      <th>hypoxia</th>\n",
       "      <th>cholesterol_homeostasis</th>\n",
       "      <th>mitotic_spindle</th>\n",
       "      <th>wnt_beta_catenin_signaling</th>\n",
       "      <th>tgf_beta_signaling</th>\n",
       "      <th>il6_jak_stat3_signaling</th>\n",
       "      <th>dna_repair</th>\n",
       "      <th>g2m_checkpoint</th>\n",
       "      <th>...</th>\n",
       "      <th>heme_metabolism</th>\n",
       "      <th>coagulation</th>\n",
       "      <th>il2_stat5_signaling</th>\n",
       "      <th>bile_acid_metabolism</th>\n",
       "      <th>peroxisome</th>\n",
       "      <th>allograft_rejection</th>\n",
       "      <th>spermatogenesis</th>\n",
       "      <th>kras_signaling_up</th>\n",
       "      <th>kras_signaling_dn</th>\n",
       "      <th>pancreas_beta_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2M</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AARS1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABAT</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4379</th>\n",
       "      <td>ZNRF4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4380</th>\n",
       "      <td>ZPBP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4381</th>\n",
       "      <td>ZW10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4382</th>\n",
       "      <td>ZWINT</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4383</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4384 rows × 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  tnfa_signaling_via_nfkb  hypoxia  cholesterol_homeostasis  \\\n",
       "0          A2M                    False    False                    False   \n",
       "1         AAAS                    False    False                    False   \n",
       "2        AADAT                    False    False                    False   \n",
       "3        AARS1                    False    False                    False   \n",
       "4         ABAT                    False    False                    False   \n",
       "...        ...                      ...      ...                      ...   \n",
       "4379     ZNRF4                    False    False                    False   \n",
       "4380      ZPBP                    False    False                    False   \n",
       "4381      ZW10                    False    False                    False   \n",
       "4382     ZWINT                    False    False                    False   \n",
       "4383       ZYX                    False    False                    False   \n",
       "\n",
       "      mitotic_spindle  wnt_beta_catenin_signaling  tgf_beta_signaling  \\\n",
       "0               False                       False               False   \n",
       "1               False                       False               False   \n",
       "2               False                       False               False   \n",
       "3               False                       False               False   \n",
       "4               False                       False               False   \n",
       "...               ...                         ...                 ...   \n",
       "4379            False                       False               False   \n",
       "4380            False                       False               False   \n",
       "4381            False                       False               False   \n",
       "4382            False                       False               False   \n",
       "4383            False                       False               False   \n",
       "\n",
       "      il6_jak_stat3_signaling  dna_repair  g2m_checkpoint  ...  \\\n",
       "0                        True       False           False  ...   \n",
       "1                       False        True           False  ...   \n",
       "2                       False       False           False  ...   \n",
       "3                       False       False           False  ...   \n",
       "4                       False       False           False  ...   \n",
       "...                       ...         ...             ...  ...   \n",
       "4379                    False       False           False  ...   \n",
       "4380                    False       False           False  ...   \n",
       "4381                    False       False           False  ...   \n",
       "4382                    False        True           False  ...   \n",
       "4383                    False       False           False  ...   \n",
       "\n",
       "      heme_metabolism  coagulation  il2_stat5_signaling  bile_acid_metabolism  \\\n",
       "0               False         True                False                 False   \n",
       "1               False        False                False                 False   \n",
       "2               False        False                False                 False   \n",
       "3               False        False                False                 False   \n",
       "4               False        False                False                 False   \n",
       "...               ...          ...                  ...                   ...   \n",
       "4379            False        False                False                 False   \n",
       "4380            False        False                False                 False   \n",
       "4381            False        False                False                 False   \n",
       "4382            False        False                False                 False   \n",
       "4383            False        False                False                 False   \n",
       "\n",
       "      peroxisome  allograft_rejection  spermatogenesis  kras_signaling_up  \\\n",
       "0          False                False            False              False   \n",
       "1          False                False            False              False   \n",
       "2          False                False            False              False   \n",
       "3          False                 True            False              False   \n",
       "4          False                False            False              False   \n",
       "...          ...                  ...              ...                ...   \n",
       "4379       False                False             True              False   \n",
       "4380       False                False             True              False   \n",
       "4381       False                False            False              False   \n",
       "4382       False                False            False              False   \n",
       "4383       False                False            False              False   \n",
       "\n",
       "      kras_signaling_dn  pancreas_beta_cells  \n",
       "0                 False                False  \n",
       "1                 False                False  \n",
       "2                 False                False  \n",
       "3                 False                False  \n",
       "4                 False                False  \n",
       "...                 ...                  ...  \n",
       "4379              False                False  \n",
       "4380              False                False  \n",
       "4381              False                False  \n",
       "4382              False                False  \n",
       "4383              False                False  \n",
       "\n",
       "[4384 rows x 51 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "# Load the JSON data\n",
    "with open(\"data/gsea/h.all.v2023.2.Hs.json\") as file:\n",
    "    gsea_data = json.load(file)\n",
    "\n",
    "# Create a dictionary to store the data\n",
    "data_dict = {}\n",
    "\n",
    "# Iterate over each hallmark property\n",
    "for hallmark, hallmark_data in gsea_data.items():\n",
    "    # Extract the gene symbols for the current hallmark\n",
    "    gene_symbols = hallmark_data['geneSymbols']\n",
    "    \n",
    "    # Add the gene symbols as a column in the data dictionary\n",
    "    data_dict[hallmark] = pd.Series(True, index=gene_symbols)\n",
    "\n",
    "# Create the DataFrame from the data dictionary\n",
    "gsea = pd.DataFrame(data_dict)\n",
    "\n",
    "# Fill missing values with False\n",
    "gsea.fillna(False, inplace=True)\n",
    "\n",
    "# Reset the index to make gene IDs as rows\n",
    "gsea.reset_index(inplace=True)\n",
    "gsea.rename(columns={'index': 'gene_name'}, inplace=True)\n",
    "\n",
    "# Function to remove \"HALLMARK_\" prefix and convert to lowercase\n",
    "def clean_column_name(col_name):\n",
    "    if col_name.startswith(\"HALLMARK_\"):\n",
    "        return col_name[len(\"HALLMARK_\"):].lower()\n",
    "    return col_name.lower()\n",
    "\n",
    "# Apply the function to all column names\n",
    "gsea.rename(columns=lambda col: clean_column_name(col), inplace=True)\n",
    "\n",
    "gsea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name                            0\n",
       "tnfa_signaling_via_nfkb              0\n",
       "hypoxia                              0\n",
       "cholesterol_homeostasis              0\n",
       "mitotic_spindle                      0\n",
       "wnt_beta_catenin_signaling           0\n",
       "tgf_beta_signaling                   0\n",
       "il6_jak_stat3_signaling              0\n",
       "dna_repair                           0\n",
       "g2m_checkpoint                       0\n",
       "apoptosis                            0\n",
       "notch_signaling                      0\n",
       "adipogenesis                         0\n",
       "estrogen_response_early              0\n",
       "estrogen_response_late               0\n",
       "androgen_response                    0\n",
       "myogenesis                           0\n",
       "protein_secretion                    0\n",
       "interferon_alpha_response            0\n",
       "interferon_gamma_response            0\n",
       "apical_junction                      0\n",
       "apical_surface                       0\n",
       "hedgehog_signaling                   0\n",
       "complement                           0\n",
       "unfolded_protein_response            0\n",
       "pi3k_akt_mtor_signaling              0\n",
       "mtorc1_signaling                     0\n",
       "e2f_targets                          0\n",
       "myc_targets_v1                       0\n",
       "myc_targets_v2                       0\n",
       "epithelial_mesenchymal_transition    0\n",
       "inflammatory_response                0\n",
       "xenobiotic_metabolism                0\n",
       "fatty_acid_metabolism                0\n",
       "oxidative_phosphorylation            0\n",
       "glycolysis                           0\n",
       "reactive_oxygen_species_pathway      0\n",
       "p53_pathway                          0\n",
       "uv_response_up                       0\n",
       "uv_response_dn                       0\n",
       "angiogenesis                         0\n",
       "heme_metabolism                      0\n",
       "coagulation                          0\n",
       "il2_stat5_signaling                  0\n",
       "bile_acid_metabolism                 0\n",
       "peroxisome                           0\n",
       "allograft_rejection                  0\n",
       "spermatogenesis                      0\n",
       "kras_signaling_up                    0\n",
       "kras_signaling_dn                    0\n",
       "pancreas_beta_cells                  0\n",
       "gene_id                              2\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_dict = dict(zip(genes.gene_name, genes.gene_id))\n",
    "\n",
    "gsea[\"gene_id\"] = gsea[\"gene_name\"].map(genes_dict)\n",
    "\n",
    "gsea.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manually taken from ENSEMBL\n",
    "gsea.loc[gsea.gene_name == 'METTL7B', 'gene_id'] = 'ENSG00000163575'\n",
    "gsea.loc[gsea.gene_name == 'THEG', 'gene_id'] = 'ENSG00000105549'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# intogen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>in_cgc</th>\n",
       "      <th>is_driver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCC4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACKR3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACSL3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name  in_cgc  is_driver\n",
       "0     ABCC4   False      False\n",
       "1      ABL1    True      False\n",
       "2      ABL2    True      False\n",
       "4     ACKR3    True      False\n",
       "5     ACSL3    True      False"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = [\"SYMBOL\", \"CGC_GENE\", \"CGC_CANCER_GENE\"]\n",
    "\n",
    "into = pd.read_csv(\"data/intogen/Compendium_Cancer_Genes.tsv\", sep=\"\\t\", usecols=cols)\n",
    "into.drop_duplicates(subset=[\"SYMBOL\"], inplace=True)\n",
    "\n",
    "into.rename(columns={\"SYMBOL\": \"gene_name\", \"CGC_GENE\": \"in_cgc\", \"CGC_CANCER_GENE\": \"is_driver\"}, inplace=True)\n",
    "\n",
    "\n",
    "into.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "genes_set = set(genes.gene_name)\n",
    "into_set = set(into.gene_name)\n",
    "\n",
    "# see if all genes in into are in genes\n",
    "print(genes_set.issuperset(into_set))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_14504/923762337.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  genes = genes.merge(into, on=\"gene_name\", how=\"left\").fillna({\"in_cgc\": False, \"is_driver\": False})\n"
     ]
    }
   ],
   "source": [
    "# add driver data to genes table\n",
    "genes = genes.merge(into, on=\"gene_name\", how=\"left\").fillna({\"in_cgc\": False, \"is_driver\": False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>in_cgc</th>\n",
       "      <th>is_driver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>DDX11L17</td>\n",
       "      <td>1</td>\n",
       "      <td>182696</td>\n",
       "      <td>184174</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000228037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>2581560</td>\n",
       "      <td>2584533</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000284616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5301928</td>\n",
       "      <td>5307394</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000157911</td>\n",
       "      <td>PEX10</td>\n",
       "      <td>1</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63236</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KI270711.1</td>\n",
       "      <td>4612</td>\n",
       "      <td>29626</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63237</th>\n",
       "      <td>ENSG00000275987</td>\n",
       "      <td>U1</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>30437</td>\n",
       "      <td>30580</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63238</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>35407</td>\n",
       "      <td>35916</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63239</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>31698</td>\n",
       "      <td>32528</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63240</th>\n",
       "      <td>ENSG00000275405</td>\n",
       "      <td>U1</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63241 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id gene_name         chr    start      end  \\\n",
       "0      ENSG00000279928  DDX11L17           1   182696   184174   \n",
       "1      ENSG00000228037       NaN           1  2581560  2584533   \n",
       "2      ENSG00000142611    PRDM16           1  3069168  3438621   \n",
       "3      ENSG00000284616       NaN           1  5301928  5307394   \n",
       "4      ENSG00000157911     PEX10           1  2403964  2413797   \n",
       "...                ...       ...         ...      ...      ...   \n",
       "63236  ENSG00000271254       NaN  KI270711.1     4612    29626   \n",
       "63237  ENSG00000275987        U1  KI270713.1    30437    30580   \n",
       "63238  ENSG00000268674       NaN  KI270713.1    35407    35916   \n",
       "63239  ENSG00000277475       NaN  KI270713.1    31698    32528   \n",
       "63240  ENSG00000275405        U1  KI270713.1    21861    22024   \n",
       "\n",
       "                 gene_biotype  in_cgc  is_driver  \n",
       "0      unprocessed_pseudogene   False      False  \n",
       "1                      lncRNA   False      False  \n",
       "2              protein_coding    True      False  \n",
       "3                      lncRNA   False      False  \n",
       "4              protein_coding   False      False  \n",
       "...                       ...     ...        ...  \n",
       "63236          protein_coding   False      False  \n",
       "63237                   snRNA   False      False  \n",
       "63238          protein_coding   False      False  \n",
       "63239          protein_coding   False      False  \n",
       "63240                   snRNA   False      False  \n",
       "\n",
       "[63241 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_biotype\n",
       "protein_coding    87\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[genes.is_driver].gene_biotype.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SAVE TO DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4384"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_df.to_sql('mirnas', engine, if_exists='replace', index=False)\n",
    "genes.to_sql('genes', engine, if_exists='replace', index=False)\n",
    "transcripts.to_sql('transcripts', engine, if_exists='replace', index=False)\n",
    "gsea.to_sql('gsea', engine, if_exists='replace', index=False)\n",
    " "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirscribe-vcf-8VJm4tPg-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
