{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/nazif/thesis/mirscribe-vcf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nazif/.cache/pypoetry/virtualenvs/mirscribe-vcf-8VJm4tPg-py3.10/lib/python3.10/site-packages/IPython/core/magics/osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.\n",
      "  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"
     ]
    }
   ],
   "source": [
    "cd ../.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import create_engine\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# db engine\n",
    "engine = create_engine('sqlite:///data/db/mirscribe.db')\n",
    "\n",
    "# pyensembl db engine\n",
    "pyensembl = create_engine('sqlite:////home/nazif/.cache/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.gtf.db')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "      <td>14363</td>\n",
       "      <td>29806</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>1</td>\n",
       "      <td>52473</td>\n",
       "      <td>54936</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id   gene_name chr  start    end gene_biotype\n",
       "0  ENSG00000223972     DDX11L1   1  11869  14412   pseudogene\n",
       "1  ENSG00000227232      WASH7P   1  14363  29806   pseudogene\n",
       "2  ENSG00000243485  MIR1302-10   1  29554  31109      lincRNA\n",
       "3  ENSG00000237613     FAM138A   1  34554  36081      lincRNA\n",
       "4  ENSG00000268020      OR4G4P   1  52473  54936   pseudogene"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_sql(\"gene\", pyensembl, columns=[\"gene_id\", \"gene_name\", \"seqname\", \"start\", \"end\", \"gene_biotype\"])\n",
    "\n",
    "genes = genes.replace(r'^\\s*$', np.nan, regex=True)\n",
    "genes = genes.infer_objects(copy=False)\n",
    "\n",
    "genes.rename(columns={\"seqname\": \"chr\"}, inplace=True)\n",
    "\n",
    "genes.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# add biomart data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "e75 = pd.read_csv(\"data/biomart/ensembl75_g37.tsv\", sep=\"\\t\")\n",
    "colnames = {'Gene stable ID': 'gene_id',\n",
    "            'EntrezGene ID': 'entrez_id',\n",
    "            'Gene description': 'gene_description',\n",
    "            'Gene name': 'gene_name',\n",
    "            'Gene type': 'biomart_biotype',\n",
    "            }\n",
    "\n",
    "e75 = e75.rename(columns={\"Ensembl Gene ID\": \"gene_id\"})\n",
    "\n",
    "\n",
    "e112 = pd.read_csv(\"data/biomart/ensembl112_g37.tsv\", sep=\"\\t\")\n",
    "colnames = {'Gene stable ID': 'gene_id',\n",
    "            'NCBI gene (formerly Entrezgene) ID': 'entrez_id',\n",
    "            'Gene description': 'gene_description',\n",
    "            'Gene type': 'biomart_biotype',\n",
    "            }\n",
    "\n",
    "e112 = e112.rename(columns=colnames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "there are 0 genes in pyensembl75 not in e75\n",
      "there are 425 genes in e75 not in pyensembl75\n",
      "there are 63677 genes in both\n",
      "######################\n",
      "there are 0 genes in pyensembl75 not in e112\n",
      "there are 0 genes in e112 not in pyensembl75\n",
      "there are 63677 genes in both\n"
     ]
    }
   ],
   "source": [
    "pyen = set(genes.gene_id.tolist())\n",
    "e75s = set(e75.gene_id.tolist())\n",
    "e112s = set(e112.gene_id.tolist())\n",
    "\n",
    "# compare \n",
    "print(f\"there are {len(pyen - e75s)} genes in pyensembl75 not in e75\")\n",
    "print(f\"there are {len(e75s - pyen)} genes in e75 not in pyensembl75\")\n",
    "print(f\"there are {len(pyen & e75s)} genes in both\")\n",
    "print(\"######################\")\n",
    "print(f\"there are {len(pyen - e112s)} genes in pyensembl75 not in e112\")\n",
    "print(f\"there are {len(e112s - pyen)} genes in e112 not in pyensembl75\")\n",
    "print(f\"there are {len(pyen & e112s)} genes in both\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "whole pipeline is done with pyensembl75, therefore i must use biomart export from ensembl 112. They're fully compatible"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "entrez_id\n",
       "80864.0        20\n",
       "259215.0       17\n",
       "100169763.0    16\n",
       "554313.0       16\n",
       "8367.0         16\n",
       "               ..\n",
       "149837.0        1\n",
       "200634.0        1\n",
       "6718.0          1\n",
       "29125.0         1\n",
       "100653067.0     1\n",
       "Name: count, Length: 25788, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e112.entrez_id.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>biomart_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24600</th>\n",
       "      <td>ENSG00000206329</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25563</th>\n",
       "      <td>ENSG00000254875</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26024</th>\n",
       "      <td>ENSG00000244444</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>EGF-like-domain, multiple 8 [Source:HGNC Symbo...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30170</th>\n",
       "      <td>ENSG00000206256</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33847</th>\n",
       "      <td>ENSG00000242038</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>EGF-like-domain, multiple 8 [Source:HGNC Symbo...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36844</th>\n",
       "      <td>ENSG00000227600</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37573</th>\n",
       "      <td>ENSG00000239974</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>EGF-like-domain, multiple 8 [Source:HGNC Symbo...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39416</th>\n",
       "      <td>ENSG00000236649</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39637</th>\n",
       "      <td>ENSG00000240389</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>EGF-like-domain, multiple 8 [Source:HGNC Symbo...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53506</th>\n",
       "      <td>ENSG00000168452</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53508</th>\n",
       "      <td>ENSG00000168452</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53510</th>\n",
       "      <td>ENSG00000168452</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53663</th>\n",
       "      <td>ENSG00000243897</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>EGF-like-domain, multiple 8 [Source:HGNC Symbo...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53664</th>\n",
       "      <td>ENSG00000243897</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>EGF-like-domain, multiple 8 [Source:HGNC Symbo...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53665</th>\n",
       "      <td>ENSG00000243897</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>EGF-like-domain, multiple 8 [Source:HGNC Symbo...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55606</th>\n",
       "      <td>ENSG00000221988</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56161</th>\n",
       "      <td>ENSG00000241404</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>EGF-like-domain, multiple 8 [Source:HGNC Symbo...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60857</th>\n",
       "      <td>ENSG00000228116</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61058</th>\n",
       "      <td>ENSG00000231618</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>palmitoyl-protein thioesterase 2 [Source:HGNC ...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61287</th>\n",
       "      <td>ENSG00000240592</td>\n",
       "      <td>80864.0</td>\n",
       "      <td>EGF-like-domain, multiple 8 [Source:HGNC Symbo...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id  entrez_id  \\\n",
       "24600  ENSG00000206329    80864.0   \n",
       "25563  ENSG00000254875    80864.0   \n",
       "26024  ENSG00000244444    80864.0   \n",
       "30170  ENSG00000206256    80864.0   \n",
       "33847  ENSG00000242038    80864.0   \n",
       "36844  ENSG00000227600    80864.0   \n",
       "37573  ENSG00000239974    80864.0   \n",
       "39416  ENSG00000236649    80864.0   \n",
       "39637  ENSG00000240389    80864.0   \n",
       "53506  ENSG00000168452    80864.0   \n",
       "53508  ENSG00000168452    80864.0   \n",
       "53510  ENSG00000168452    80864.0   \n",
       "53663  ENSG00000243897    80864.0   \n",
       "53664  ENSG00000243897    80864.0   \n",
       "53665  ENSG00000243897    80864.0   \n",
       "55606  ENSG00000221988    80864.0   \n",
       "56161  ENSG00000241404    80864.0   \n",
       "60857  ENSG00000228116    80864.0   \n",
       "61058  ENSG00000231618    80864.0   \n",
       "61287  ENSG00000240592    80864.0   \n",
       "\n",
       "                                        gene_description biomart_biotype  \n",
       "24600  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "25563                                                NaN  protein_coding  \n",
       "26024  EGF-like-domain, multiple 8 [Source:HGNC Symbo...  protein_coding  \n",
       "30170  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "33847  EGF-like-domain, multiple 8 [Source:HGNC Symbo...  protein_coding  \n",
       "36844  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "37573  EGF-like-domain, multiple 8 [Source:HGNC Symbo...  protein_coding  \n",
       "39416  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "39637  EGF-like-domain, multiple 8 [Source:HGNC Symbo...  protein_coding  \n",
       "53506  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "53508  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "53510  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "53663  EGF-like-domain, multiple 8 [Source:HGNC Symbo...  protein_coding  \n",
       "53664  EGF-like-domain, multiple 8 [Source:HGNC Symbo...  protein_coding  \n",
       "53665  EGF-like-domain, multiple 8 [Source:HGNC Symbo...  protein_coding  \n",
       "55606  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "56161  EGF-like-domain, multiple 8 [Source:HGNC Symbo...  protein_coding  \n",
       "60857  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "61058  palmitoyl-protein thioesterase 2 [Source:HGNC ...  protein_coding  \n",
       "61287  EGF-like-domain, multiple 8 [Source:HGNC Symbo...  protein_coding  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e112[e112.entrez_id == 80864.0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 l...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "      <td>14363</td>\n",
       "      <td>29806</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>WAS protein family homolog 7 pseudogene [Sourc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>microRNA 1302-10 [Source:HGNC Symbol;Acc:38233]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>family with sequence similarity 138, member A ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>1</td>\n",
       "      <td>52473</td>\n",
       "      <td>54936</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>olfactory receptor, family 4, subfamily G, mem...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id   gene_name chr  start    end gene_biotype  \\\n",
       "0  ENSG00000223972     DDX11L1   1  11869  14412   pseudogene   \n",
       "1  ENSG00000227232      WASH7P   1  14363  29806   pseudogene   \n",
       "2  ENSG00000243485  MIR1302-10   1  29554  31109      lincRNA   \n",
       "3  ENSG00000237613     FAM138A   1  34554  36081      lincRNA   \n",
       "4  ENSG00000268020      OR4G4P   1  52473  54936   pseudogene   \n",
       "\n",
       "                                    gene_description  \n",
       "0  DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 l...  \n",
       "1  WAS protein family homolog 7 pseudogene [Sourc...  \n",
       "2    microRNA 1302-10 [Source:HGNC Symbol;Acc:38233]  \n",
       "3  family with sequence similarity 138, member A ...  \n",
       "4  olfactory receptor, family 4, subfamily G, mem...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e112_to_merge = e112[[\"gene_id\",\"gene_description\"]].drop_duplicates()\n",
    "df = pd.merge(genes, e112_to_merge, on=\"gene_id\", how=\"left\")\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id                 0\n",
       "gene_name               0\n",
       "chr                     0\n",
       "start                   0\n",
       "end                     0\n",
       "gene_biotype            0\n",
       "gene_description    23066\n",
       "dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"gene_description\"] = df[\"gene_description\"].fillna(\"no_description\")\n",
    "# df[\"entrez_id\"] = df[\"entrez_id\"].fillna(-1).astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id             0\n",
       "gene_name           0\n",
       "chr                 0\n",
       "start               0\n",
       "end                 0\n",
       "gene_biotype        0\n",
       "gene_description    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# adding oncokb data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>ENSG00000097007</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>ENSG00000142208</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALK</td>\n",
       "      <td>ENSG00000171094</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMER1</td>\n",
       "      <td>ENSG00000184675</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APC</td>\n",
       "      <td>ENSG00000134982</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name          gene_id  is_oncogene  is_tumor_suppressor  \\\n",
       "0      ABL1  ENSG00000097007         True                False   \n",
       "1      AKT1  ENSG00000142208         True                False   \n",
       "2       ALK  ENSG00000171094         True                False   \n",
       "3     AMER1  ENSG00000184675        False                 True   \n",
       "4       APC  ENSG00000134982        False                 True   \n",
       "\n",
       "   is_cosmic_tier_1  \n",
       "0              True  \n",
       "1              True  \n",
       "2              True  \n",
       "3              True  \n",
       "4              True  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kb = pd.read_csv(\"data/oncokb/oncokb.csv\")\n",
    "kb = kb.rename(columns={\"gene_symbol\": \"gene_name\", \"ensembl_gene_id\": \"gene_id\", \"is_oncogene\": \"is_oncogene\", \"is_tumor_suppressor\": \"is_tumor_suppressor\", \"is_cosmic_tier_1\": \"is_cosmic_tier_1\"})\n",
    "\n",
    "# drop columns where all bool values are false\n",
    "cols = [\"is_oncogene\", \"is_tumor_suppressor\", \"is_cosmic_tier_1\"]\n",
    "kb = kb[~(kb[cols] == False).all(axis=1)]\n",
    "\n",
    "kb.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>GATA1</td>\n",
       "      <td>ENSG00000102145</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>H3C2</td>\n",
       "      <td>ENSG00000286522</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>IKZF1</td>\n",
       "      <td>ENSG00000185811</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>NSD3</td>\n",
       "      <td>ENSG00000147548</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>PIM1</td>\n",
       "      <td>ENSG00000137193</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1128</th>\n",
       "      <td>WDCP</td>\n",
       "      <td>ENSG00000163026</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1130</th>\n",
       "      <td>WRN</td>\n",
       "      <td>ENSG00000165392</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1132</th>\n",
       "      <td>XPA</td>\n",
       "      <td>ENSG00000136936</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1135</th>\n",
       "      <td>YWHAE</td>\n",
       "      <td>ENSG00000108953</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1142</th>\n",
       "      <td>ZNF331</td>\n",
       "      <td>ENSG00000130844</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>213 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name          gene_id  is_oncogene  is_tumor_suppressor  \\\n",
       "143      GATA1  ENSG00000102145        False                False   \n",
       "146       H3C2  ENSG00000286522        False                False   \n",
       "148      IKZF1  ENSG00000185811        False                False   \n",
       "169       NSD3  ENSG00000147548        False                False   \n",
       "177       PIM1  ENSG00000137193        False                False   \n",
       "...        ...              ...          ...                  ...   \n",
       "1128      WDCP  ENSG00000163026        False                False   \n",
       "1130       WRN  ENSG00000165392        False                False   \n",
       "1132       XPA  ENSG00000136936        False                False   \n",
       "1135     YWHAE  ENSG00000108953        False                False   \n",
       "1142    ZNF331  ENSG00000130844        False                False   \n",
       "\n",
       "      is_cosmic_tier_1  \n",
       "143               True  \n",
       "146               True  \n",
       "148               True  \n",
       "169               True  \n",
       "177               True  \n",
       "...                ...  \n",
       "1128              True  \n",
       "1130              True  \n",
       "1132              True  \n",
       "1135              True  \n",
       "1142              True  \n",
       "\n",
       "[213 rows x 5 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kb[kb.is_cosmic_tier_1 & ~((kb.is_oncogene) | (kb.is_tumor_suppressor))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>IGH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name gene_id  is_oncogene  is_tumor_suppressor  is_cosmic_tier_1\n",
       "715       IGH     NaN        False                False              True"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kb[kb.gene_name == \"IGH\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28581</th>\n",
       "      <td>ENSG00000254017</td>\n",
       "      <td>IGHEP2</td>\n",
       "      <td>9</td>\n",
       "      <td>5113549</td>\n",
       "      <td>5114804</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42364</th>\n",
       "      <td>ENSG00000211891</td>\n",
       "      <td>IGHE</td>\n",
       "      <td>14</td>\n",
       "      <td>106064224</td>\n",
       "      <td>106068065</td>\n",
       "      <td>IG_C_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42374</th>\n",
       "      <td>ENSG00000253692</td>\n",
       "      <td>IGHEP1</td>\n",
       "      <td>14</td>\n",
       "      <td>106185955</td>\n",
       "      <td>106188864</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61064</th>\n",
       "      <td>ENSG00000270340</td>\n",
       "      <td>IGHE</td>\n",
       "      <td>HG1592_PATCH</td>\n",
       "      <td>106064224</td>\n",
       "      <td>106068065</td>\n",
       "      <td>IG_C_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61074</th>\n",
       "      <td>ENSG00000270715</td>\n",
       "      <td>IGHEP1</td>\n",
       "      <td>HG1592_PATCH</td>\n",
       "      <td>106185955</td>\n",
       "      <td>106188864</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id gene_name           chr      start        end  \\\n",
       "28581  ENSG00000254017    IGHEP2             9    5113549    5114804   \n",
       "42364  ENSG00000211891      IGHE            14  106064224  106068065   \n",
       "42374  ENSG00000253692    IGHEP1            14  106185955  106188864   \n",
       "61064  ENSG00000270340      IGHE  HG1592_PATCH  106064224  106068065   \n",
       "61074  ENSG00000270715    IGHEP1  HG1592_PATCH  106185955  106188864   \n",
       "\n",
       "          gene_biotype  \n",
       "28581  IG_C_pseudogene  \n",
       "42364        IG_C_gene  \n",
       "42374  IG_C_pseudogene  \n",
       "61064        IG_C_gene  \n",
       "61074  IG_C_pseudogene  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[genes.gene_name.str.startswith(\"IGHE\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54911</th>\n",
       "      <td>ENSG00000170889</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>19</td>\n",
       "      <td>54704610</td>\n",
       "      <td>54752862</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61426</th>\n",
       "      <td>ENSG00000273385</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>HSCHR19LRC_LRC_J_CTG1</td>\n",
       "      <td>54704726</td>\n",
       "      <td>54711511</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61453</th>\n",
       "      <td>ENSG00000273467</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>HSCHR19LRC_LRC_S_CTG1</td>\n",
       "      <td>54704726</td>\n",
       "      <td>54711511</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61481</th>\n",
       "      <td>ENSG00000272943</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>HSCHR19LRC_LRC_I_CTG1</td>\n",
       "      <td>54704726</td>\n",
       "      <td>54711511</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61508</th>\n",
       "      <td>ENSG00000271572</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>HG1079_PATCH</td>\n",
       "      <td>54703996</td>\n",
       "      <td>54752257</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61637</th>\n",
       "      <td>ENSG00000272643</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>HSCHR19LRC_LRC_T_CTG1</td>\n",
       "      <td>54704726</td>\n",
       "      <td>54711511</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61666</th>\n",
       "      <td>ENSG00000263076</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>HSCHR19LRC_COX1_CTG1</td>\n",
       "      <td>54704428</td>\n",
       "      <td>54752625</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61740</th>\n",
       "      <td>ENSG00000263332</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>HSCHR19LRC_PGF1_CTG1</td>\n",
       "      <td>54704601</td>\n",
       "      <td>54752858</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61883</th>\n",
       "      <td>ENSG00000272614</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>HSCHR19LRC_PGF2_CTG1</td>\n",
       "      <td>54704726</td>\n",
       "      <td>54711511</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61973</th>\n",
       "      <td>ENSG00000273135</td>\n",
       "      <td>RPS9</td>\n",
       "      <td>HSCHR19LRC_COX2_CTG1</td>\n",
       "      <td>54704726</td>\n",
       "      <td>54711511</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id gene_name                    chr     start       end  \\\n",
       "54911  ENSG00000170889      RPS9                     19  54704610  54752862   \n",
       "61426  ENSG00000273385      RPS9  HSCHR19LRC_LRC_J_CTG1  54704726  54711511   \n",
       "61453  ENSG00000273467      RPS9  HSCHR19LRC_LRC_S_CTG1  54704726  54711511   \n",
       "61481  ENSG00000272943      RPS9  HSCHR19LRC_LRC_I_CTG1  54704726  54711511   \n",
       "61508  ENSG00000271572      RPS9           HG1079_PATCH  54703996  54752257   \n",
       "61637  ENSG00000272643      RPS9  HSCHR19LRC_LRC_T_CTG1  54704726  54711511   \n",
       "61666  ENSG00000263076      RPS9   HSCHR19LRC_COX1_CTG1  54704428  54752625   \n",
       "61740  ENSG00000263332      RPS9   HSCHR19LRC_PGF1_CTG1  54704601  54752858   \n",
       "61883  ENSG00000272614      RPS9   HSCHR19LRC_PGF2_CTG1  54704726  54711511   \n",
       "61973  ENSG00000273135      RPS9   HSCHR19LRC_COX2_CTG1  54704726  54711511   \n",
       "\n",
       "         gene_biotype  \n",
       "54911  protein_coding  \n",
       "61426  protein_coding  \n",
       "61453  protein_coding  \n",
       "61481  protein_coding  \n",
       "61508  protein_coding  \n",
       "61637  protein_coding  \n",
       "61666  protein_coding  \n",
       "61740  protein_coding  \n",
       "61883  protein_coding  \n",
       "61973  protein_coding  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[genes.gene_name == \"RPS9\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '2', '3', '4', '5', '6', '7', 'X', '8', '9', '10', '11', '12',\n",
       "       '13', '14', '15', '16', '17', '18', '20', '19', 'Y', '22', '21',\n",
       "       'HG1287_PATCH', 'HG1459_PATCH', 'HSCHR6_MHC_SSTO',\n",
       "       'HSCHR6_MHC_MCF', 'HSCHR6_MHC_COX', 'HSCHR6_MHC_MANN',\n",
       "       'HSCHR6_MHC_APD', 'HSCHR6_MHC_QBL', 'HSCHR6_MHC_DBB',\n",
       "       'HG1433_PATCH', 'HG1257_PATCH', 'HG1497_PATCH',\n",
       "       'HG1436_HG1432_PATCH', 'HG1211_PATCH', 'HSCHR17_1', 'HG1292_PATCH',\n",
       "       'HSCHR5_1_CTG1', 'HG1592_PATCH', 'HG1425_PATCH', 'HG1462_PATCH',\n",
       "       'HG7_PATCH', 'HG1458_PATCH', 'HSCHR19LRC_LRC_J_CTG1',\n",
       "       'HSCHR19LRC_LRC_S_CTG1', 'HSCHR19LRC_LRC_I_CTG1', 'HG1079_PATCH',\n",
       "       'HG1443_HG1444_PATCH', 'HG979_PATCH', 'HSCHR19LRC_LRC_T_CTG1',\n",
       "       'HSCHR19LRC_COX1_CTG1', 'HSCHR19LRC_PGF1_CTG1', 'HG1453_PATCH',\n",
       "       'HG1423_PATCH', 'HG1437_PATCH', 'HG865_PATCH', 'HG1434_PATCH',\n",
       "       'HG1438_PATCH', 'HSCHR19LRC_PGF2_CTG1', 'HG1293_PATCH',\n",
       "       'HG1426_PATCH', 'HSCHR19LRC_COX2_CTG1', 'HG1308_PATCH',\n",
       "       'HG1463_PATCH', 'HG987_PATCH', 'HG19_PATCH', 'HG953_PATCH',\n",
       "       'HSCHR4_1', 'HG730_PATCH', 'HG1350_HG959_PATCH', 'GL000192.1',\n",
       "       'HG306_PATCH', 'HG1082_HG167_PATCH', 'HG75_PATCH', 'HG1424_PATCH',\n",
       "       'HG174_HG254_PATCH', 'HG357_PATCH', 'HG29_PATCH', 'HG256_PATCH',\n",
       "       'HG183_PATCH', 'HG1146_PATCH', 'HG1441_PATCH', 'HG871_PATCH',\n",
       "       'HG104_HG975_PATCH', 'HG1442_PATCH', 'HG185_PATCH', 'HG122_PATCH',\n",
       "       'HSCHR12_2_CTG2', 'HG1091_PATCH', 'HG544_PATCH', 'HG1440_PATCH',\n",
       "       'HG686_PATCH', 'HG1439_PATCH', 'HSCHR15_1_CTG8', 'HG747_PATCH',\n",
       "       'HSCHR19_1_CTG3', 'HG748_PATCH', 'HSCHR4_1_CTG6', 'HG1133_PATCH',\n",
       "       'HSCHR1_3_CTG31', 'HG115_PATCH', 'HG736_PATCH', 'HG417_PATCH',\n",
       "       'HG1591_PATCH', 'HG745_PATCH', 'HG1595_PATCH', 'HG79_PATCH',\n",
       "       'HG946_PATCH', 'HG388_HG400_PATCH', 'HG281_PATCH',\n",
       "       'HSCHR10_1_CTG5', 'HG237_PATCH', 'HG1074_PATCH', 'HG1032_PATCH',\n",
       "       'HSCHR15_1_CTG4', 'HG962_PATCH', 'HSCHR18_1_CTG1_1',\n",
       "       'HG480_HG481_PATCH', 'HG944_PATCH', 'HSCHR17_4_CTG4',\n",
       "       'HG536_PATCH', 'HG444_PATCH', 'HG27_PATCH', 'HG344_PATCH',\n",
       "       'HG1435_PATCH', 'HSCHR17_1_CTG1', 'HG1063_PATCH', 'HG957_PATCH',\n",
       "       'HG1304_PATCH', 'HG14_PATCH', 'HG706_PATCH', 'HG729_PATCH',\n",
       "       'HG348_PATCH', 'HG1699_PATCH', 'HSCHR5_3_CTG1', 'HSCHR17_2_CTG4',\n",
       "       'HG329_PATCH', 'HG385_PATCH', 'HG1502_PATCH', 'GL000225.1',\n",
       "       'HG243_PATCH', 'HSCHR21_2_CTG1_1', 'HG142_HG150_NOVEL_TEST',\n",
       "       'HG1322_PATCH', 'HSCHR18_2_CTG2', 'HG311_PATCH', 'HG883_PATCH',\n",
       "       'HSCHR16_1_CTG3_1', 'GL000194.1', 'HG151_NOVEL_TEST',\n",
       "       'HG414_PATCH', 'GL000193.1', 'HSCHR19_3_CTG3', 'GL000222.1',\n",
       "       'GL000212.1', 'HG271_PATCH', 'HSCHR12_1_CTG5', 'HG1490_PATCH',\n",
       "       'GL000195.1', 'HSCHR1_1_CTG31', 'HSCHR3_1_CTG2_1', 'GL000223.1',\n",
       "       'HG506_HG507_HG1000_PATCH', 'GL000224.1', 'HSCHR10_1_CTG2',\n",
       "       'GL000219.1', 'HG339_PATCH', 'GL000205.1', 'HSCHR5_1_CTG5',\n",
       "       'HSCHR3_1_CTG1', 'GL000215.1', 'GL000216.1', 'HG186_PATCH',\n",
       "       'HSCHR9_2_CTG35', 'HSCHR19_2_CTG3', 'HG971_PATCH', 'GL000199.1',\n",
       "       'HG1501_PATCH', 'HSCHR12_1_CTG2_1', 'HG905_PATCH',\n",
       "       'HSCHR18_1_CTG2_1', 'HG873_PATCH', 'HSCHR12_1_CTG1',\n",
       "       'HG1486_PATCH', 'GL000211.1', 'HG996_PATCH', 'HSCHR4_1_CTG12',\n",
       "       'GL000213.1', 'HG858_PATCH', 'HSCHR9_1_CTG1', 'HSCHR22_1_CTG1',\n",
       "       'GL000220.1', 'GL000218.1', 'HSCHR18_2_CTG2_1', 'GL000209.1',\n",
       "       'HG50_PATCH', 'HSCHR19_1_CTG3_1', 'GL000221.1', 'HSCHR12_3_CTG2_1',\n",
       "       'HG989_PATCH', 'HG193_PATCH', 'HSCHR2_1_CTG1', 'HSCHR12_2_CTG2_1',\n",
       "       'HSCHR17_1_CTG4', 'GL000228.1', 'HSCHR20_1_CTG1', 'HSCHR6_1_CTG5',\n",
       "       'HSCHR2_1_CTG12', 'HSCHR4_2_CTG9', 'HG305_PATCH', 'HSCHR7_1_CTG6',\n",
       "       'HSCHR21_4_CTG1_1', 'HSCHR1_2_CTG31', 'HG299_PATCH', 'GL000191.1',\n",
       "       'HG418_PATCH', 'HSCHR5_1_CTG2', 'HG325_PATCH', 'HG1208_PATCH',\n",
       "       'HSCHR22_1_CTG2', 'HSCHR2_2_CTG12', 'HG1479_PATCH',\n",
       "       'HSCHR17_3_CTG4', 'HSCHR16_2_CTG3_1', 'HSCHR17_6_CTG4',\n",
       "       'HSCHR5_2_CTG1', 'GL000204.1', 'HG280_PATCH', 'HSCHR21_3_CTG1_1',\n",
       "       'HSCHR22_2_CTG1', 'HG1488_PATCH', 'HSCHR9_1_CTG35', 'HG1487_PATCH',\n",
       "       'HG144_PATCH', 'HSCHR17_5_CTG4', 'HG1007_PATCH', 'HG991_PATCH',\n",
       "       'HG375_PATCH', 'HG998_1_PATCH', 'HSCHR9_3_CTG35', 'HG999_2_PATCH',\n",
       "       'HG999_1_PATCH', 'HG998_2_PATCH', 'GL000233.1', 'GL000237.1',\n",
       "       'HG990_PATCH', 'GL000230.1', 'HG497_PATCH', 'GL000242.1',\n",
       "       'GL000243.1', 'GL000241.1', 'GL000236.1', 'GL000240.1',\n",
       "       'GL000196.1', 'GL000247.1', 'GL000201.1', 'HG531_PATCH',\n",
       "       'GL000231.1', 'HG1472_PATCH', 'HG1500_PATCH', 'GL000229.1', 'MT'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.chr.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>28581</th>\n",
       "      <td>ENSG00000254017</td>\n",
       "      <td>IGHEP2</td>\n",
       "      <td>9</td>\n",
       "      <td>5113549</td>\n",
       "      <td>5114804</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42369</th>\n",
       "      <td>ENSG00000253755</td>\n",
       "      <td>IGHGP</td>\n",
       "      <td>14</td>\n",
       "      <td>106130970</td>\n",
       "      <td>106136180</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42374</th>\n",
       "      <td>ENSG00000253692</td>\n",
       "      <td>IGHEP1</td>\n",
       "      <td>14</td>\n",
       "      <td>106185955</td>\n",
       "      <td>106188864</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49968</th>\n",
       "      <td>ENSG00000265243</td>\n",
       "      <td>IGLJCOR18</td>\n",
       "      <td>18</td>\n",
       "      <td>3394887</td>\n",
       "      <td>3395310</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56161</th>\n",
       "      <td>ENSG00000254029</td>\n",
       "      <td>IGLC4</td>\n",
       "      <td>22</td>\n",
       "      <td>23252998</td>\n",
       "      <td>23253245</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56163</th>\n",
       "      <td>ENSG00000254030</td>\n",
       "      <td>IGLC5</td>\n",
       "      <td>22</td>\n",
       "      <td>23257805</td>\n",
       "      <td>23258097</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56165</th>\n",
       "      <td>ENSG00000222037</td>\n",
       "      <td>IGLC6</td>\n",
       "      <td>22</td>\n",
       "      <td>23261707</td>\n",
       "      <td>23262023</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56483</th>\n",
       "      <td>ENSG00000254127</td>\n",
       "      <td>IGLCOR22-1</td>\n",
       "      <td>22</td>\n",
       "      <td>32595906</td>\n",
       "      <td>32596221</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56491</th>\n",
       "      <td>ENSG00000271093</td>\n",
       "      <td>IGLCOR22-2</td>\n",
       "      <td>22</td>\n",
       "      <td>32752663</td>\n",
       "      <td>32752975</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61069</th>\n",
       "      <td>ENSG00000271411</td>\n",
       "      <td>IGHGP</td>\n",
       "      <td>HG1592_PATCH</td>\n",
       "      <td>106130970</td>\n",
       "      <td>106136180</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61074</th>\n",
       "      <td>ENSG00000270715</td>\n",
       "      <td>IGHEP1</td>\n",
       "      <td>HG1592_PATCH</td>\n",
       "      <td>106185955</td>\n",
       "      <td>106188864</td>\n",
       "      <td>IG_C_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id   gene_name           chr      start        end  \\\n",
       "28581  ENSG00000254017      IGHEP2             9    5113549    5114804   \n",
       "42369  ENSG00000253755       IGHGP            14  106130970  106136180   \n",
       "42374  ENSG00000253692      IGHEP1            14  106185955  106188864   \n",
       "49968  ENSG00000265243   IGLJCOR18            18    3394887    3395310   \n",
       "56161  ENSG00000254029       IGLC4            22   23252998   23253245   \n",
       "56163  ENSG00000254030       IGLC5            22   23257805   23258097   \n",
       "56165  ENSG00000222037       IGLC6            22   23261707   23262023   \n",
       "56483  ENSG00000254127  IGLCOR22-1            22   32595906   32596221   \n",
       "56491  ENSG00000271093  IGLCOR22-2            22   32752663   32752975   \n",
       "61069  ENSG00000271411       IGHGP  HG1592_PATCH  106130970  106136180   \n",
       "61074  ENSG00000270715      IGHEP1  HG1592_PATCH  106185955  106188864   \n",
       "\n",
       "          gene_biotype  \n",
       "28581  IG_C_pseudogene  \n",
       "42369  IG_C_pseudogene  \n",
       "42374  IG_C_pseudogene  \n",
       "49968  IG_C_pseudogene  \n",
       "56161  IG_C_pseudogene  \n",
       "56163  IG_C_pseudogene  \n",
       "56165  IG_C_pseudogene  \n",
       "56483  IG_C_pseudogene  \n",
       "56491  IG_C_pseudogene  \n",
       "61069  IG_C_pseudogene  \n",
       "61074  IG_C_pseudogene  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[genes.gene_biotype == \"IG_C_pseudogene\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id\n",
       "True     931\n",
       "False      7\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kb.gene_id.isin(df.gene_id).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "146      H3C2\n",
       "453    H2AC17\n",
       "700      H4C9\n",
       "715       IGH\n",
       "716       IGK\n",
       "717       IGL\n",
       "741     MLLT6\n",
       "Name: gene_name, dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_gene_names = kb[~kb.gene_id.isin(df.gene_id)].gene_name\n",
    "missing_gene_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "146      H3C2 is HIST1H3D in grch37\n",
    "\n",
    "453    H2AC17 is whole locus\n",
    "\n",
    "700      H4C9 not found in grch37\n",
    "\n",
    "715       IGH whole locus\n",
    "\n",
    "716       IGK whole locus\n",
    "\n",
    "717       IGL whole locus\n",
    "\n",
    "741     MLLT6 found in gene_names\n",
    "\n",
    "whole loci are discarded. other genes are appended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 l...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "      <td>14363</td>\n",
       "      <td>29806</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>WAS protein family homolog 7 pseudogene [Sourc...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>microRNA 1302-10 [Source:HGNC Symbol;Acc:38233]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>family with sequence similarity 138, member A ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>1</td>\n",
       "      <td>52473</td>\n",
       "      <td>54936</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>olfactory receptor, family 4, subfamily G, mem...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id   gene_name chr  start    end gene_biotype  \\\n",
       "0  ENSG00000223972     DDX11L1   1  11869  14412   pseudogene   \n",
       "1  ENSG00000227232      WASH7P   1  14363  29806   pseudogene   \n",
       "2  ENSG00000243485  MIR1302-10   1  29554  31109      lincRNA   \n",
       "3  ENSG00000237613     FAM138A   1  34554  36081      lincRNA   \n",
       "4  ENSG00000268020      OR4G4P   1  52473  54936   pseudogene   \n",
       "\n",
       "                                    gene_description is_oncogene  \\\n",
       "0  DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 l...         NaN   \n",
       "1  WAS protein family homolog 7 pseudogene [Sourc...         NaN   \n",
       "2    microRNA 1302-10 [Source:HGNC Symbol;Acc:38233]         NaN   \n",
       "3  family with sequence similarity 138, member A ...         NaN   \n",
       "4  olfactory receptor, family 4, subfamily G, mem...         NaN   \n",
       "\n",
       "  is_tumor_suppressor is_cosmic_tier_1  \n",
       "0                 NaN              NaN  \n",
       "1                 NaN              NaN  \n",
       "2                 NaN              NaN  \n",
       "3                 NaN              NaN  \n",
       "4                 NaN              NaN  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kb = kb.drop(columns=[\"gene_name\"])\n",
    "df = pd.merge(df, kb, on=\"gene_id\", how=\"left\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# adding details of 2 genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18369</th>\n",
       "      <td>ENSG00000197409</td>\n",
       "      <td>HIST1H3D</td>\n",
       "      <td>6</td>\n",
       "      <td>26197068</td>\n",
       "      <td>26199521</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>histone cluster 1, H3d [Source:HGNC Symbol;Acc...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48365</th>\n",
       "      <td>ENSG00000108292</td>\n",
       "      <td>MLLT6</td>\n",
       "      <td>17</td>\n",
       "      <td>36861795</td>\n",
       "      <td>36886056</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>myeloid/lymphoid or mixed-lineage leukemia (tr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id gene_name chr     start       end    gene_biotype  \\\n",
       "18369  ENSG00000197409  HIST1H3D   6  26197068  26199521  protein_coding   \n",
       "48365  ENSG00000108292     MLLT6  17  36861795  36886056  protein_coding   \n",
       "\n",
       "                                        gene_description is_oncogene  \\\n",
       "18369  histone cluster 1, H3d [Source:HGNC Symbol;Acc...         NaN   \n",
       "48365  myeloid/lymphoid or mixed-lineage leukemia (tr...         NaN   \n",
       "\n",
       "      is_tumor_suppressor is_cosmic_tier_1  \n",
       "18369                 NaN              NaN  \n",
       "48365                 NaN              NaN  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.gene_name.isin([\"MLLT6\", \"HIST1H3D\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df.gene_name == 'HIST1H3D', 'is_cosmic_tier_1'] = True\n",
    "df.loc[df.gene_name == 'MLLT6', 'is_cosmic_tier_1'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id                    0\n",
       "gene_name                  0\n",
       "chr                        0\n",
       "start                      0\n",
       "end                        0\n",
       "gene_biotype               0\n",
       "gene_description           0\n",
       "is_oncogene            62746\n",
       "is_tumor_suppressor    62746\n",
       "is_cosmic_tier_1       62744\n",
       "dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_8486/809247794.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df.fillna(False, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 l...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "      <td>14363</td>\n",
       "      <td>29806</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>WAS protein family homolog 7 pseudogene [Sourc...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>microRNA 1302-10 [Source:HGNC Symbol;Acc:38233]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>family with sequence similarity 138, member A ...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>1</td>\n",
       "      <td>52473</td>\n",
       "      <td>54936</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>olfactory receptor, family 4, subfamily G, mem...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id   gene_name chr  start    end gene_biotype  \\\n",
       "0  ENSG00000223972     DDX11L1   1  11869  14412   pseudogene   \n",
       "1  ENSG00000227232      WASH7P   1  14363  29806   pseudogene   \n",
       "2  ENSG00000243485  MIR1302-10   1  29554  31109      lincRNA   \n",
       "3  ENSG00000237613     FAM138A   1  34554  36081      lincRNA   \n",
       "4  ENSG00000268020      OR4G4P   1  52473  54936   pseudogene   \n",
       "\n",
       "                                    gene_description  is_oncogene  \\\n",
       "0  DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 l...        False   \n",
       "1  WAS protein family homolog 7 pseudogene [Sourc...        False   \n",
       "2    microRNA 1302-10 [Source:HGNC Symbol;Acc:38233]        False   \n",
       "3  family with sequence similarity 138, member A ...        False   \n",
       "4  olfactory receptor, family 4, subfamily G, mem...        False   \n",
       "\n",
       "   is_tumor_suppressor  is_cosmic_tier_1  \n",
       "0                False             False  \n",
       "1                False             False  \n",
       "2                False             False  \n",
       "3                False             False  \n",
       "4                False             False  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.fillna(False, inplace=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# adding intogen data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>is_driver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACVR1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARAF</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASXL2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BAP1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name  is_driver\n",
       "0     ACVR1       True\n",
       "1    ACVR2A       True\n",
       "2      ARAF       True\n",
       "3     ASXL2       True\n",
       "4      BAP1       True"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "into = pd.read_json(\"data/intogen/driver_genes.json\")\n",
    "\n",
    "into.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name\n",
       "True     86\n",
       "False     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "into.gene_name.isin(df.gene_name).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "78    TENT5C\n",
       "Name: gene_name, dtype: object"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "into[~into.gene_name.isin(df.gene_name)].gene_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "FAM46C is the grch37 name of TENT5C\n",
    "\n",
    "source: https://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000183508;r=1:117606048-117628389;t=ENST00000369448"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.merge(df, into, on=\"gene_name\", how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df.gene_name == \"FAM46C\", \"is_driver\"] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_8486/2844565991.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df[\"is_driver\"] = df[\"is_driver\"].fillna(False)\n"
     ]
    }
   ],
   "source": [
    "df[\"is_driver\"] = df[\"is_driver\"].fillna(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "      <th>is_driver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 l...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "      <td>14363</td>\n",
       "      <td>29806</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>WAS protein family homolog 7 pseudogene [Sourc...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>microRNA 1302-10 [Source:HGNC Symbol;Acc:38233]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>1</td>\n",
       "      <td>34554</td>\n",
       "      <td>36081</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>family with sequence similarity 138, member A ...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>1</td>\n",
       "      <td>52473</td>\n",
       "      <td>54936</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>olfactory receptor, family 4, subfamily G, mem...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63672</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>MT</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mitochondrially encoded NADH dehydrogenase 6 [...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63673</th>\n",
       "      <td>ENSG00000210194</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>MT</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mitochondrially encoded tRNA glutamic acid [So...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63674</th>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>MT</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>mitochondrially encoded cytochrome b [Source:H...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63675</th>\n",
       "      <td>ENSG00000210195</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>MT</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mitochondrially encoded tRNA threonine [Source...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63676</th>\n",
       "      <td>ENSG00000210196</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>MT</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mitochondrially encoded tRNA proline [Source:H...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63677 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id   gene_name chr  start    end    gene_biotype  \\\n",
       "0      ENSG00000223972     DDX11L1   1  11869  14412      pseudogene   \n",
       "1      ENSG00000227232      WASH7P   1  14363  29806      pseudogene   \n",
       "2      ENSG00000243485  MIR1302-10   1  29554  31109         lincRNA   \n",
       "3      ENSG00000237613     FAM138A   1  34554  36081         lincRNA   \n",
       "4      ENSG00000268020      OR4G4P   1  52473  54936      pseudogene   \n",
       "...                ...         ...  ..    ...    ...             ...   \n",
       "63672  ENSG00000198695      MT-ND6  MT  14149  14673  protein_coding   \n",
       "63673  ENSG00000210194       MT-TE  MT  14674  14742         Mt_tRNA   \n",
       "63674  ENSG00000198727      MT-CYB  MT  14747  15887  protein_coding   \n",
       "63675  ENSG00000210195       MT-TT  MT  15888  15953         Mt_tRNA   \n",
       "63676  ENSG00000210196       MT-TP  MT  15956  16023         Mt_tRNA   \n",
       "\n",
       "                                        gene_description  is_oncogene  \\\n",
       "0      DEAD/H (Asp-Glu-Ala-Asp/His) box helicase 11 l...        False   \n",
       "1      WAS protein family homolog 7 pseudogene [Sourc...        False   \n",
       "2        microRNA 1302-10 [Source:HGNC Symbol;Acc:38233]        False   \n",
       "3      family with sequence similarity 138, member A ...        False   \n",
       "4      olfactory receptor, family 4, subfamily G, mem...        False   \n",
       "...                                                  ...          ...   \n",
       "63672  mitochondrially encoded NADH dehydrogenase 6 [...        False   \n",
       "63673  mitochondrially encoded tRNA glutamic acid [So...        False   \n",
       "63674  mitochondrially encoded cytochrome b [Source:H...        False   \n",
       "63675  mitochondrially encoded tRNA threonine [Source...        False   \n",
       "63676  mitochondrially encoded tRNA proline [Source:H...        False   \n",
       "\n",
       "       is_tumor_suppressor  is_cosmic_tier_1  is_driver  \n",
       "0                    False             False      False  \n",
       "1                    False             False      False  \n",
       "2                    False             False      False  \n",
       "3                    False             False      False  \n",
       "4                    False             False      False  \n",
       "...                    ...               ...        ...  \n",
       "63672                False             False      False  \n",
       "63673                False             False      False  \n",
       "63674                False             False      False  \n",
       "63675                False             False      False  \n",
       "63676                False             False      False  \n",
       "\n",
       "[63677 rows x 11 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id                0\n",
       "gene_name              0\n",
       "chr                    0\n",
       "start                  0\n",
       "end                    0\n",
       "gene_biotype           0\n",
       "gene_description       0\n",
       "is_oncogene            0\n",
       "is_tumor_suppressor    0\n",
       "is_cosmic_tier_1       0\n",
       "is_driver              0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from sqlalchemy import create_engine\n",
    "\n",
    "# # db engine\n",
    "# engine = create_engine('sqlite:///data/db/mirscribe.db')\n",
    "\n",
    "# df.to_sql(name=\"genes\", con=engine, if_exists=\"replace\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>is_oncogene</th>\n",
       "      <th>is_tumor_suppressor</th>\n",
       "      <th>is_cosmic_tier_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000097007</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000142208</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000171094</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000184675</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000134982</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1138</th>\n",
       "      <td>ENSG00000181722</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1139</th>\n",
       "      <td>ENSG00000185650</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1140</th>\n",
       "      <td>ENSG00000152518</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1142</th>\n",
       "      <td>ENSG00000130844</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1143</th>\n",
       "      <td>ENSG00000141579</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>938 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id  is_oncogene  is_tumor_suppressor  is_cosmic_tier_1\n",
       "0     ENSG00000097007         True                False              True\n",
       "1     ENSG00000142208         True                False              True\n",
       "2     ENSG00000171094         True                False              True\n",
       "3     ENSG00000184675        False                 True              True\n",
       "4     ENSG00000134982        False                 True              True\n",
       "...               ...          ...                  ...               ...\n",
       "1138  ENSG00000181722         True                 True             False\n",
       "1139  ENSG00000185650        False                 True             False\n",
       "1140  ENSG00000152518        False                 True             False\n",
       "1142  ENSG00000130844        False                False              True\n",
       "1143  ENSG00000141579        False                 True             False\n",
       "\n",
       "[938 rows x 4 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "414"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kb[kb.is_oncogene].gene_id.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "576"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kb[kb.is_cosmic_tier_1].gene_id.unique())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirscribe-vcf-8VJm4tPg-py3.10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
